\documentclass[letterpaper,11pt]{amsart}

\usepackage{amsthm,amsmath,amsfonts,amssymb}
\usepackage[left=1.0in, right=1.0in, vmargin=1in]{geometry}
\usepackage{pgfplots}
\usepackage{setspace}
\usepackage[normalem]{ulem}
\usepackage{graphicx}
\usepackage{cite}
\usepackage{algorithmic}
%\usepackage[usenames,dvipsnames]{color}
\usepackage{url}
\usepackage{xspace}

\usepackage{subfigure}

\usepackage[footnotesize]{caption}

\renewcommand{\algorithmicrequire}{\textbf{Input:}}
\renewcommand{\algorithmicforall}{\textbf{for each}}

\def\eps{\varepsilon}

\def\prefix{\mathop{\rm prefix}}
\def\suffix{\mathop{\rm suffix}}

\newcommand{\isempty}{\textsc{isempty}}
\newcommand{\isprefix}{\textsc{isprefix}}
\newcommand{\valid}{\textsc{valid}}
\newcommand{\vstart}{\textsc{start}}
\newcommand{\vend}{\textsc{end}}
\newcommand{\indeg}{\textsc{indeg}}
\newcommand{\outdeg}{\textsc{outdeg}}
\newcommand{\edge}{\textsc{edge}}
\newcommand{\offset}{\textsc{offset}}
\newcommand{\leftmer}{\textsc{left}}
\newcommand{\rightmer}{\textsc{right}}

\newcommand{\ecoli}{E.~\textit{coli}\xspace}


%\newtheoremstyle{theorem}{\topsep}{\topsep}
\newtheorem{Th}{Theorem}
\newtheorem{Lemma}[Th]{Lemma}
\newtheorem{Corol}[Th]{Corollary}

\newtheorem{Def}{Definition}
\newtheorem{Problem}{Problem}
\newtheorem{Remark}{Remark}


\begin{document}
\def\ra#1{\rotatebox{90}{\parbox{1.5in}{#1}}}
\def\mrk#1{{\bf #1}}


\newlength{\commentlength}
\setlength{\commentlength}{1in}
\newcommand{\son}[1]{\marginpar{\colorbox{blue!15!white}{\parbox{\commentlength}{\scriptsize #1}}}}
\newcommand{\glenn}[1]{\marginpar{\colorbox{red!20!white}{\parbox{\commentlength}{\scriptsize #1}}}}
\newcommand{\maxal}[1]{\marginpar{\colorbox{blue!40!white}{\parbox{\commentlength}{\scriptsize #1}}}}
\newcommand{\pavel}[1]{\marginpar{\colorbox{green!40!white}{\parbox{\commentlength}{\scriptsize #1}}}}


\title{Jumping libraries for SPADES - For local lab's use only}
\maketitle
\section{Motivations}
 
In recent publications, we used pair-end  and paired-end reads interchangeably. In fact, they are very different; we only worked with paired-end reads that have insert size from 200 to 600 bp. 
Mate-pairs reads have longer insert size,
ranging from 1000 bp to 65,000 bp (and can even be longer). Surprisingly, until now, only ALLPATH-LG is able to demonstrate the ability of  
assembling complicated genomes from only these cheap and high-thoughput sequencing data. April 2011, ALLPATH-LG announced that it can assemble human genomes only from these short reads and obtained assemblies with
N50 = 11.5 Mb and 
base accuracy 99.95\%. All other 
existing assemblers produce N50 ranging from several Kb to several hundreds of Kb using these high throughput short reads data with various insert size libraries. 

Integrating of mate-pairs (with long insert size) and normal paired-end reads into SPADES will improve the assemblies greatly. Below we describe some properties of mate-paired. 

\section{Properties of mate-pairs from Illumina platform}

\subsection*{Insert size as a bimodal distribution}
The insert size of mate-pairs reads generated from Illuminate mate-pair protocol actually contains both mate-pairs and paired-end reads (For details see Illumina Mate Pairs Protocol).
Figure~\ref{fig:bimodal} shows the insert size distribution of an Illumina mate-paired library for \emph{Pichia stipitis} illustrates two peaks in the distribution.
The one with IS $\approx$ 300 corresponds to the normal paired-end reads,
    the one with peak $\approx$ 3300 corresponds to  the mate-paired  reads.

   \begin{figure}
\includegraphics[width=60mm]{fig/stipitis.png}
\caption{Andrey Prjibelski -- The insert size distribution of an Illumina mate-paired library for \emph{Pichia stipitis} illustrates two peaks in the distribution. The one with IS $\approx$ 300 corresponds to the normal paired-end reads,
    the one with peak $\approx$ 3300 corresponds to  the mate-paired  reads. }
\label{fig:bimodal}
\end{figure}


\subsection*{Direction of mate-pairs and filtering contaminated paired-end reads}
The problem is that given a pair of read $(r_1,r_2)$, how to recognize if it is actually a mate-pair or a contaminated pair-end reads. One way to filter these unwanted paired-end reads is to build the de Bruijn graph 
on the reads from the ordinary paired-end data and check if $r_1$ and $r_2$ are in the approximated distance of $300$ with the matching the direction. Note that mate-pairs have out-ward facing reads while
paired-ends have inward facing reads.
\subsection*{Changeover errors}

Mate-pair reads also have problem of changing over within the reads. In other words, a read contain sequence from one and the other end and create chimeric reads. This can be corrected by mapping reads to the de Bruijn graph.

\subsection*{Coverage}

The increase of the insert size is sacrificed by the decrease of the coverage. For large insert size, (65 kb), the coverage may decrease from 10 to 100 folds. 
\subsection*{Data} Multiple mate-pair libs generated from Illumina technology are available. There may be other similar platforms that eliminate the problems from Illumina protocol but we could not find any data available.

\subsection*{Single Cell}
Mate-pairs with long insert size for single cell have been generated, and they are up to 9 Kb. 

\section{Multiple libraries}
Obviously, using mate-pairs with long insert size requires at least one normal paired-ends. Low coverage and Multiple libraries are challenges to repeat resolvers.
\bibliographystyle{plain}
\bibliography{mybib}


\end{document}

