configfile: "config.yaml"
GAPATH = config["spalignerpath"]
GACFG = config["spalignercfg"]
GRAPH = config["graph"] 
K = config["kmer"]

ALL_ORGS=["ecoli"] # ["ecoli", "celegans", "scerevisiae"]
ALL_READS=["realpb2000", "realnp2000", "simpb2000", "simnp2000"]

SCRIPT_PATH=os.getcwd()

workdir: config["datapath"]

rule all:
    input:
        expand("{org}/SPAligner/output/aln_{reads}.tsv", reads=ALL_READS, org=ALL_ORGS)


rule run_aligner:
    output:
        tsv = "{org}/SPAligner/output/aln_{reads}.tsv"
    log:
        "{org}/SPAligner/output/aln_{reads}.log"
    benchmark:
        repeat("{org}/SPAligner/benchmark/align_{reads}.tsv", 3)
    threads: 16
    run:
        if wildcards.reads.endswith("pb2000"):
            shell("{GAPATH} {SCRIPT_PATH}/{GACFG} --spades -g {wildcards.org}/{GRAPH} -s {wildcards.org}/input/{wildcards.reads}.fasta -K {K} -d pacbio -o {wildcards.org}/SPAligner/output/aln_{wildcards.reads} > {log}")
        else:
            shell("{GAPATH} {SCRIPT_PATH}/{GACFG} --spades -g {wildcards.org}/{GRAPH} -s {wildcards.org}/input/{wildcards.reads}.fasta -K {K} -d nanopore -o {wildcards.org}/SPAligner/output/aln_{wildcards.reads} > {log}")