<html>
<head>
    <title>dipSPAdes 1.0 Manual</title>
    <style type="text/css">
        .code {
            background-color: lightgray;
        }
    </style>
</head>

<body>
<h1>dipSPAdes 1.0 manual</h1>

<ol>
	<li><a href = "#ds">What is dipSPAdes?</a></li>
	<ol>
		<li><a href = "#ds_pipeline">dipSPAdes pipeline</a></li>
	</ol>
	<li><a href = "#sec_install">Installing dipSPAdes</a></li>

	<li><a href = "#running">Running dipSPAdes</a></li>
		<ol>
			<li><a href = "#ds_input">dipSPAdes input</a></li>
			<li><a href = "#options">dipSPAdes command line options</a></li>
			<ol>
				<li><a href = "#basic_opt">Basic options</a></li>
				<li><a href = "#input_data">Input data</a></li>
				<li><a href = "#adv_opt">Advanced options</a></li>
				<li><a href = "#examples">Examples</a></li>
                <li><a href = "#adv_examples">Examples of advanced options usage</a></li>
			</ol>

			<li><a href = "#ds_output">dipSPAdes output</a></li>
			<ol>
                <li><a href = "#align_output">Haplocontigs alignment output</a></li>
				<li><a href = "#hapl_assembly_output">Haplotype assembly output</a></li>
			</ol>
		</ol>

	<li><a href = "#feedback">Feedback and bug reports</a></li>
</ol>

<a id = "ds"></a>
<h2>1. What is dipSPAdes?</h2>
dipSPAdes is a genome assembler designed specifically for diploid highly polymorphic genomes based on SPAdes.
It takes advantage of divergence between haplomes in repetitive genome regions to resolve them and construct longer contigs. 
dipSPAdes produces consensus contigs (representing a consensus of both haplomes for the orthologous regions) and performs haplotype assembly.
Note that dipSPAdes can only benefit from high polymorphism rate (at least 0.4%).
For the data with low polymorphism rate no improvement in terms of N50 is expected. 

<a id = "ds_pipeline"></a>
<h3>dipSPAdes pipeline</h3>
dipSPAdes Pipeline consists of three steps.
<ol>
    <li>Assembly of haplocontigs (contigs representing both haplomes)</li>
    <li>Consensus contigs construction</li>
    <li>Haplotype assembly</li>
</ol>


<a id = sec_install></a>
<h2>2. Installing dipSPAdes</h2>
dipSPAdes comes as a part of SPAdes assembler package.
See <a href = manual.html#sec2>SPAdes manual</a> for installation instructions.

You can additionally verify your dipSPAdes installation by running
<pre class="code">
<code>
    &lt;spades installation dir>/dipspades.py --test
</code>
</pre>

<a id = "running"></a>
<h2>3. Running dipSPAdes</h2>
<a id = "ds_input"></a>
<h3>2.1 dipSPAdes input</h2>
dipSPAdes takes as an input libraries of reads in the same way as described in <a href=manual.html#sec3.1>SPAdes manual</a>.
Also dipSPAdes can use haplocontigs provided by user.
For example one can assemble haplocontigs with another assembler and run dipSPAdes with these haplocontigs (instead of reads) as an input.
In this case the first step of dipSPAdes pipeline is skipped.
If both reads and haplocontigs are provided as an input, dipSPAdes first computes haplocontigs from reads and then uses mixture of computed haplocontigs and haplocontigs provided by user for further steps.
<a id = "options"></a>
<h3>2.2 dipSPAdes command line options </h2>
To run dipSPAdes from the command line, type<br>
<pre class="code">
<code>
dipspades.py [options] -o &lt;output_dir>
</code>
</pre><br>
Note that we assume that SPAdes installation directory is added to the <code>PATH</code> variable (provide full path to dipSPAdes executable otherwise: <code>&lt;spades installation dir>/dipspades.py</code>).

<a id = "basic_opt"></a>
<h4>2.2.1 Basic options</h4>
<p>
    <code>-o &lt;output_dir> </code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Specify the output directory. Required option.
</p>

<p>
    <code>--test</code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Runs SPAdes on the toy data set; see <a href="#sec_install">section 2</a>.
</p>

<p>
    <code>-h</code> (or <code>--help</code>)<br>
    &nbsp;&nbsp;&nbsp;&nbsp;Prints help.
    
<a id = "input_data"></a>
<h4>2.2.2 Input data</h4>
For input reads specfication use options of SPAdes described in <a href=manual.html#inputdata>SPAdes manual</a>.

<p>
    <code>--hap &lt;file_name></code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;File with haplocontigs in FASTA format. Note that dipSPAdes can use any number of haplocontig files.
</p>

<a id = "adv_opt"></a>
<h4>2.2.3 Advanced options</h4>
<p>
    <code>--expect-gaps</code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Use to indicate that gaps in genome coverage are to be expected in significant amounts (e.g. for datasets with relatively low coverage).
</p>

<p>
    <code>--expect-rearrangements</code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Use to indicate extreme heterozygosity rate in haplomes (e.g. haplomes differ by long insertions/delitions).
</p>

<a id = "examples"></a>
<h4>2.2.4 Examples</h4>
To perform assembly (construct consensus contigs and perform haplotype assembly) of diploid genome from paired-end reads (<code>reads_left.fastq</code> and <code>reads_right.fastq</code>):
<pre class="code">
<code>
dipspades.py -1 reads_left.fastq -2 reads_right.fastq -o output_dir
</code>
</pre></br>

To perform assembly (construct consensus contigs and perform haplotype assembly) of diploid genome from preliminary computed haplocontigs (<code>haplocontigs1.fasta</code> and <code>haplocontigs2.fasta</code>):
<pre class="code">
<code>
dipspades.py --hap haplocontigs1.fasta --hap haplocontigs2.fasta -o output_dir
</code>
</pre></br>

To construct haplocontigs from paired-end reads (<code>reads_left.fastq</code> and <code>reads_right.fastq</code>),
then mix constructed haplocontigs with contigs from <code>haplocontigs.fasta</code> and use this mixture for further steps (consensus contigs construction and haplotype assembly):
<pre class="code">
<code>
dipspades.py -1 reads_left.fastq -2 reads_right.fastq --hap haplocontigs.fasta -o output_dir
</code>
</pre></br>

<a id = "adv_examples"></a>
<h4>Examples of advanced options usage</h4>
To perform assembly of diploid genome with additional options:
<pre class="code">
<code>
dipspades.py -1 reads_left.fastq -2 reads_right.fastq --allow-gaps -o output_dir
</code>
</pre></br>

To relaunch steps 2 and 3 of dipSPAdes (see <a href=#ds_pipeline>dipSPAdes pipeline section</a>) with different set of advanced options you can use haplocontigs constructed in the previous run (see <a href=#ds_output>dipSPAdes output section</a>):
<pre class="code">
<code>
dipspades.py -hap output_dir/haplocontigs.fasta --allow-gaps --expect-rearrangements -o new_output_dir
</code>
</pre></br>

<a id = "ds_output"></a>
<h3>2.3 dipSPAdes output</h3>
dipSPAdes produces the following output:<br>
<ul>
	<li><b>haplocontigs.fasta</b> - file in FASTA format. with computed haplocontigs (if input reads were provided). </li>
	<li><b>consensus_contigs.fasta</b> - file in FASTA format with a set of constructed consensus contigs </li>
	<li><b>paired_consensus_contigs.fasta</b> - file in FASTA format with a subset of consensus contigs that have a polymorphism detected on them. </li>
	<li><b>unpaired_consensus_contigs.fasta</b> - file in FASTA format with a subset of consensus contigs that have no polymorphism detected on them. These contigs are potentially redundant.</li>
	<li><b>haplocontigs_alignment.out</b> - file with recorded haplocontigs that correspond to homologous regions on haplomes.</li>
	<li><b>haplotype_assembly.out</b> - result of haplotype assembly </li> 
	<li><b>conservative_regions.fasta</b> - file in FASTA format with conservative regions of diploid genome </li>   
	<li><b>possibly_conservative_regions.fasta</b> - file in FASTA format with unresolved regions of haplocontigs that may be either conservative or repetitive. </li>
</ul>  

<a id="align_output"></a>
<h4>2.3.1 Haplocontigs alignment output</h4>
File <b>haplocontigs_alignment.out</b> contains blocks the following structure:<br>
<pre class = "code">
<code>
Consensus contig: CONSENSUS_CONTIG_NAME
    Overlapped haplocontigs:
        HAPLOCONTIG_NAME_1 HAPLOCONTIG_NAME_2
                         ...
    Aligned pairs:
        HAPLOCONTIG_NAME_3 HAPLOCONTIG_NAME_4
                        ...
</code>
</pre>
Each block corresponds to alignment of haplocontigs to concrete consensus contigs. Name of consensus contig, <code>CONSENSUS_CONTIG_NAME</code>, coincides with the name in file <b>consensus_contigs.fasta</b>. Subblock "Overlapped contigs" contains pairs of names of haplocontigs that overlaped and result in consensus contig. Subblock "Aligned pairs" contains pairs of names of haplocontigs that correspond to homologous regions of haplomes.  

<a id="hapl_assembly_output"></a>
<h4>2.3.2 Haplotype assembly output</h4>
File <b>haplotype_assembly.out</b> contains lines with the following structure:<br>
<pre class = "code">
<code>
HAPLOCONTIG_NAME_1	HAPLOCONTIG_NAME_2
</code>
</pre>
where HAPLOCONTIG_NAME_1 and HAPLOCONTIG_NAME_2 are names of homologous haplocontigs which correspond to different haplomes. Names correspond to names of haplocontigs specified as input using options <code>--hap</code> or computed at the first step.

<a id = "feedback"></a>
<h2>Feedback and bug reports</h2>
Your comments, bug reports, and suggestions are very welcomed. <br> 
If you have trouble running dipSPAdes, please provide us with the files params.txt and dipspades.log from the directory &lt; output_dir>. <br>
Address for communications: <a href="mailto:spades.support@bioinf.spbau.ru" target="_blank">spades.support@bioinf.spbau.ru</a>.

</body>
</html>
