<html>
<head>
    <title>SPAdes 2.1.0 Manual</title>
    <style type="text/css">
        .code {
            background-color: lightgray;
        }
    </style>
</head>
<body>
<h1>SPAdes 2.1.0 Manual</h1>

<p>SPAdes stands for St. Petersburg genome assembler. It is intended for both single cell MDA and standard
(multi-cell) assemblies. This manual will help you to install and run SPAdes. You can find the latest SPAdes release
at <a href="http://bioinf.spbau.ru/spades">http://bioinf.spbau.ru/spades</a>. The latest version of this manual can be found <a href="http://spades.bioinf.spbau.ru/release2.1.0/manual.html">here</a>.
</p>


<h2>1. Installation</h2>

<p>
    SPAdes requires a 64-bit Linux system. Correcting errors in our E. coli datasets requires about 70 Gb of RAM.
    Assembling our test the <a href="http://spades.bioinf.spbau.ru/spades_test_datasets/E.coli/reads/ECOLI_SC/">single cell  E. coli dataset</a> with SPAdes requires 6 Gb of RAM.
    For the <a href="http://spades.bioinf.spbau.ru/spades_test_datasets/E.coli/reads/ECOLI_MC/">multi-cell E. coli dataset</a> it will be about 700 Mb of RAM. Also SPAdes requires
    Python 2.4 or higher installed.
</p>

<p>If you have root privileges and you want to install SPAdes for all users you can install Debian package as described in section 1.3 or RPM package as described in section 1.4. It will allow users
to run <code>spades.py</code> from any directory without <code>./</code> prefix.</p>

<h3>1.1 SPAdes tar.gz file</h3>

<p>You can download <a href="http://spades.bioinf.spbau.ru/release2.1.0/spades_2.1.0.tar.gz">SPAdes tar.gz</a> and extract it:
</p>

<pre  class="code">
<code>
    wget http://spades.bioinf.spbau.ru/release2.1.0/spades_2.1.0.tar.gz
    tar -xzf spades_2.1.0.tar.gz
    cd spades_2.1.0
</code>
</pre>

<p>
    SPAdes depends on the following libraries for compiling its code:
</p>

    <ul>
        <li>g++ (version 4.4 or higher)</li>
        <li>cmake (version 2.6 or higher)</li>
        <li>boost (version 1.42 or higher)</li>
        <li>zlib</li>
    </ul>

<p>If you are not able to match these requirements please use downloaded binaries as described in the next section.</p>

<h3>1.2 Downloading binaries</h3>

<p>If you want to run SPAdes with default parameters you need to run the following scripts: </p>

<pre class="code">
<code>
    ./spades_download_binary.py 21 33 55
    ./spades_download_bayeshammer.py
</code>
</pre>

<p>It will download binaries to <code>.spades/release2.1.0</code> directory in your user home.
SPAdes uses binaries optimized for each k value (see section  2.3) and you can download binaries
for the required k values with <code>./spades_download_binary.py</code> script using space-separated values of k as an argument.</p>

<h3>1.3 Installing Debian package</h3>

<p>First, add the repository containing SPAdes by inserting the following line to the end of the file
    <code>/etc/apt/sources.list</code>:
</p>

<pre class="code">
<code>
    deb http://debian.bioinf.spbau.ru /
</code>
</pre>

<p>Note that the space before the last slash is required. Update the package list by typing</p>

<pre class="code">
<code>
    sudo apt-get update
</code>
</pre>

<p>After that, SPAdes can be installed just by typing</p>


<pre class="code">
<code>
    sudo apt-get install spades
</code>
</pre>

<h3>1.4 Installing RPM package</h3>

<p>Your can download the RPM package <a href="http://spades.bioinf.spbau.ru/release2.1.0/spades-2.1.0.rpm">here</a>. The RPM installation procedure varies on different systems, but on
    <a href="http://fedoraproject.org/">Fedora</a> we recommend
</p>

<pre class="code">
<code>
    sudo yum install spades-2.1.0.rpm
</code>
</pre>

<p>
    Please note, that some package managers will not install dependencies described in section 1.1 automatically. Anyway you can download binaries as described in section 1.2.
</p>

<h3>1.5 Testing your installation</h3>

<p>
For testing purposes, SPAdes comes with a toy dataset (first 1000 bp of E. coli). If you <code>spades.py</code> without parameters it will process this dataset
and if the installation is successful you will see something like this at the end of the log:
</p>

<pre class="code">
<code>
 * Corrected reads are in spades_output/ECOLI_1K/corrected/
 * Assembled contigs are spades_output/ECOLI_1K/spades_04.18_17.59.30/ECOLI_1K.fasta

Thank you for using SPAdes!

======= SPAdes pipeline finished
</code>
</pre>

<h2>2 Running SPAdes</h2>

<p>
There are two ways to run SPAdes for you data set you can provide the configuration file with <code>.info</code> extension
as the only <code>spades.py</code> parameter or specify the parameters from the command line.
</p>

<h3>2.2 Preparing configuration files</h3>

<p>
You can find the configuration file <code>spades_config.info</code> for the toy dataset from section 1.5 in the directory you extracted the archive (see sections 2.1 and 2.2),
if SPAdes was installed from package this file located at <code>/usr/share/spades/spades_config.info</code>.
You can use this file as the template for you datasets.
Below we list the content n example of a config file and then explain its contents in detail. In config files,
the part of a line starting with a semicolon is a comment.
</p>

<pre class="code">
<code>
; all paths are relative to the working directory but you can also use $cfg variable
; to set paths relative to this config
project_name ECOLI_1K ; optional. Name of this config file is used if not specified
output_dir ./spades_output

dataset
{
    paired_reads $cfg/test_dataset/ecoli_1K_1.fq.gz $cfg/test_dataset/ecoli_1K_2.fq.gz
    single_reads ; optional
    single_cell false
}

error_correction
{
    tmp_dir ; optional. &lt;output_dir>/corrected/tmp is used if not specified
    qvoffset ; optional. Either 33 or 64
    max_iterations 2
    max_threads 16
    max_memory 250 ; in GB
    gzip_output true ; optional
}

assembly
{
    iterative_K 21 33 55
    generate_sam_files true ; optional. Default is false
    gap_closer true ; optional. Default is true
}

</code>
</pre>

<p>
There are two global variables: <code>project_name</code> specifies the name of the project and <code>output_dir</code>
specifies the output directory. It is recommended to set a meaningful name for each
separate project.
</p>

<p>
The first section in the config file is the required <code>dataset</code> section that describes the dataset.
The other sections <code>error_correction</code>, <code>assembly</code>, and <code>quality_assessment</code> correspond
to three stages of SPAdes pipeline. To run only some of these three stages just remove (or
comment or rename) unnecessary stages. If you run SPAdes with the same output_dir and project_name
as one of previous runs, it checks if you have already made the error correction stage and asks if you want
to correct the reads again.
</p>

<h3>2.2.1 Dataset settings</h3>
<p>
SPAdes accepts single reads as well as forward-reverse paired end reads in FASTA and FASTQ format;
however, in order to run error correction, reads should be in FASTQ format. All files may be compressed
with gzip. At present, SPAdes can accept only one paired-end library as input.
</p>

<p>
<code>paired_reads</code> specifies input paired end reads organized into one or two files.
    <ul>
        <li>If using a single file, interlace the corresponding mated "1" and "2" reads. Put the read from the Illumina
            file <code>s_n_1_export.txt</code> first, followed by the read from the Illumina file <code>s_n_2_export.txt</code>.</li>
        <li>If using a pair of files, the first file should have the Illumina "1" subreads and the second file
            should have the Illumina "2" subreads, with corresponding reads in the exact same order. The
            two filenames should be separated by a space.</li>
    </ul>
</p>

<p>
<code>single_reads</code> specifies reads without pairs. You can specify any number of files with single reads. File
names should be separated by spaces.
</p>

<p>
<code>single_cell</code> flag is set if input data was obtained with MDA (single cell) technology.
</p>

<h3>2.2.2 Error correction settings</h3>
<p>
The <code>error_correction</code> section is responsible for correcting errors in the input reads; error correction in the
SPAdes pipeline is done with <a href="http://bioinf.spbau.ru/en/spades/bayeshammer">BayesHammer</a>,
a recently developed error correction tool that works well for the case of non-uniform coverage (e.g., on single-cell data). Note that for an accurate assembly it is important
to use this stage, especially for single-cell datasets.
</p>

<p>
<code>tmp_dir</code> specifies the directory where temporary files are stored. Note that BayesHammer uses a significant
    amount of disk space (even though it cleans up after itself).
</p>

<p>
    <code>qvoffset</code> sets PHRED quality offset in the input reads (33 or 64). If not given, BayesHammer tries to find
    this value automatically.
</p>

<p>
    <code>max_iterations</code> sets the maximum number of iterations for the error correction procedure. Usually, the
    second iteration does improve upon the first, but then there is almost no change.
</p>

<p>
    <code>max_threads</code> sets the maximum number of threads.
</p>

<p>
<code>max_memory</code> sets the maximum amount of available RAM (in Gb). BayesHammer (and the entire SPAdes
    pipeline) will fail if it attempts to allocate more RAM than specified here.
</p>

<p>
    <code>gzip_output</code> flag forces BayesHammer to output corrected reads compressed with gzip.
</p>

<h3>2.2.3 Assembly settings</h3>

<p>
    The <code>assembly</code> section controls the assembly process.
</p>

<p>
    <code>iterative_K</code> specifies one or more odd values for k-mer (vertex) sizes. Informally, smaller values of k make
    the graph more connected, but at the same time more tangled, while higher values of k may defragment
    the graph, but allow to resolve short repeats. The typical value for this variable is <code>21 33 55</code>. See the
    paper [1] for more details. Note that in the configuration file, k is the size of a vertex, while in the
    paper, k is the size of an edge (i.e., larger by one).
</p>

<p>
<code>paired_mode</code> turns the repeat resolver on/off if parameter set to true/false.
</p>

<p>
    <code>generate_sam_files</code> flag forces SPAdes to generate a SAM-file showing how the paired reads specified in
    dataset section of config file are aligned to resulting contigs. Alignments from this SAM-file can be
    visualized, e.g., by <a href="http://bioinf.scri.ac.uk/tablet/">Tablet</a> [3].
</p>

<p>
    gap_closer turns the gap closer on/off. This module restores k-mers that are not covered not by whole
    reads but by overlapping parts of two reads and are supported by paired reads. This is efficient when
    a library has a small insert size (up to 3x read length). If insert size is large, you can turn it off to
    slightly decrease the runing time of SPAdes. Note: you can decrease the running time of SPAdes by
    setting just one value of k (say, 55) and turning the gap closer on (since the gap closer usually gives
    similar results as running with several smaller values of k).
</p>


<h2>3 SPAdes Output</h2>
    Results can be found in the directory specified by <code>output_dir</code> parameter, the default value is
    <code>spades_output</code>. For each dataset SPAdes creates separate directory, the name of this folder can be set using
    <code>project_name</code> parameter. Corrected reads are stored in <code>corrected</code> subdirectory in <code>*.fastq.qz</code> files.
    The resulting contigs are stored in <code>contigs</code> subdirectory in <code>*.fasta</code> file.



<h2>4 Postprocessing</h2>
<p>
    While SPAdes produces accurate assemblies, we recommend running <a href="http://bix.ucsd.edu/SEQuel/">SEQuel</a> [4]
    after SPAdes to further  reduce the number of small errors (single nucleotide substitutions and small indels).
    Note that one can set the flag generate_sam to SPAdes (see section 2.2.3)
    to generate a SAM-file showing how the reads are aligned to the resulting contigs. This file can then be
    given to SEQuel as input.
</p>
<p>
    The current version of SPAdes does not have a scaffolding stage. One can use a separate scaffolder such as
    <a href="http://sourceforge.net/projects/operasf/">Opera</a> [2].
</p>

<h2>5 Feedback and bug reports</h2>
    We will be thankful if you help us make SPAdes better by sending your comments, bug reports, and suggestions
    to <a href="mailto:spades.support@bioinf.spbau.ru">spades.support@bioinf.spbau.ru</a>.


<h2>References</h2>
<p>
    [1] A. Bankevich, S. Nurk, et al. <a href="http://online.liebertpub.com/doi/abs/10.1089/cmb.2012.0021">SPAdes: A New Genome Assembly Algorithm and Its Applications to Single-Cell Sequencing.</a>
    Journal of Computational Biology. May 2012, 19(5): 455-477. doi:10.1089/cmb.2012.0021.
</p>
<p>
    [2] Song Gao, Wing-Kin Sung, and Niranjan Nagarajan. <a href="http://online.liebertpub.com/doi/abs/10.1089/cmb.2011.0170">Opera: reconstructing optimal genomic scaffolds with high-throughput paired-end sequences.</a>
    Journal of Computational Biology. November 2011, 18(11): 1681-1691. doi:10.1089/cmb.2011.0170.
</p>
<p>
    [3] I. Milne, M. Bayer, L. Cardle, P. Shaw, G. Stephen, F.Wright, and D. Marshall. <a href="http://bioinformatics.oxfordjournals.org/content/26/3/401.abstract">Tablet—next generation
    sequence assembly visualization.</a> Bioinformatics (2010) 26 (3): 401-402.   doi: 10.1093/bioinformatics/btp666
</p>
<p>
    [4] R. Ronen, C. Boucher, H. Chitsaz, and P. Pevzner. SEQuel: Improving the accuracy of genome assemblies.
    Bioinformatics, 2012. To appear.
</p>



</body>
</html>
