<html>
<head>
    <title>SPAdes 2.3.0 Manual</title>
    <style type="text/css">
        .code {
            background-color: lightgray;
        }
    </style>
</head>
<body>
<h1>SPAdes 2.3.0 Manual</h1>

<p>SPAdes stands for St. Petersburg genome assembler. It is intended for both standard isolates and single-cell MDA
    bacteria  assemblies. This manual will help you to install and run SPAdes. You can find the latest SPAdes release
at <a href="http://bioinf.spbau.ru/spades">http://bioinf.spbau.ru/spades</a>. The latest version of this manual can be found <a href="http://spades.bioinf.spbau.ru/release2.3.0/manual.html">here</a>.
    SPAdes version 2.3.0 released under  GPLv2 on 2 September 2012.
</p>


<h2>1. Installation</h2>

<p>
    SPAdes requires a 64-bit Linux system.
    We have two test datasets: <a href="http://spades.bioinf.spbau.ru/spades_test_datasets/ecoli_sc/">single-cell  E. coli dataset</a> and
    <a href="http://spades.bioinf.spbau.ru/spades_test_datasets/ecoli_mc/">E. coli isolates dataset</a>. SPAdes requires 35 Gb of RAM for processing these datasets. Also SPAdes requires  Python 2.4 or higher installed.
</p>

<h3>1.1 SPAdes tar.gz file</h3>

<p>To download <a href="http://spades.bioinf.spbau.ru/release2.3.0/spades_2.3.0.tar.gz">SPAdes tar.gz</a> and extract it:
</p>

<pre  class="code">
<code>
    wget http://spades.bioinf.spbau.ru/release2.3.0/spades_2.3.0.tar.gz
    tar -xzf spades_2.3.0.tar.gz
    cd spades-2.3.0
</code>
</pre>

<h3>1.2 Getting SPAdes binaries</h3>

    <p>There are two ways to obtain SPAdes binaries: download static builds from our server or compile SPAdes on you server.</p>

    <p>We recommend to download binaries with the following script:</p>

<pre class="code">
<code>
    ./spades_download_binary.py
</code>
</pre>

<p>
    Also you can compile SPAdes yourself, but SPAdes depends on the following libraries for compiling its code:
</p>

    <ul>
        <li>g++ (version 4.4 or higher)</li>
        <li>cmake (version 2.8 or higher)</li>
        <li>zlib</li>
    </ul>

<p>
    If you meet this requirements you can build SPAdes with the following script:
</p>

<pre class="code">
<code>
    ./spades_compile.sh
</code>
</pre>

<p>In both cases you should get <code>bin</code> directory with files <code>hammer</code> (error correcting module) ans <code>spades</code> (assembly module).
</p>

<a name="sec1.3"></a>
<h3>1.3 Testing your installation</h3>

<p>
For testing purposes, SPAdes comes with a toy dataset (first 1000 bp of E. coli). If you run <code>spades.py</code> with the parameter <code>--test</code>
</p>

<pre class="code">
<code>
    ./spades.py --test
</code>
</pre>

<p>
it will process this dataset and if the installation is successful you will see something like this at the end of the log:
</p>

<pre class="code">
<code>
 * Corrected reads are in spades_test/corrected/
 * Assembled contigs are spades_test/contigs.fasta

Thank you for using SPAdes!

======= SPAdes pipeline finished
</code>
</pre>

<h2>2 Running SPAdes</h2>

<h3>2.1 Input data</h3>
<p>
    SPAdes accepts single reads as well as forward-reverse paired end reads in FASTA and FASTQ format;
    however, in order to run error correction, reads should be in FASTQ format. All files may be compressed
    with gzip. At present, SPAdes can accept only one paired-end library as an input.
</p>

<p>
    SPAdes supports paired end reads organized in two separate files or combined in one:

<ul>

    <li>If the forward and reverse reads for each lane are split into separate FASTQ files (e.g.
        <code>s_1_1_sequence.txt</code> and <code>s_1_2_sequence.txt</code> for the forward and reverse reads, respectively, from lane 1),
        the second file  should have corresponding reads in the exact same order.
    </li>
    <li>
        If using a single file containing reads in both orientations, interlace the corresponding reads such that each read from
        <code>s_1_1_sequence.txt</code> is followed by the corresponding read from <code>s_1_2_sequence.txt</code>.
    </li>


</ul>

<h3>2.2 SPAdes pipeline</h3>

<p>SPAdes stores all the resulting files in the directory <code>&lt;output_dir></code>.
</p>

<p>Before starting assembly SPAdes initiates <a href="http://bioinf.spbau.ru/en/spades/bayeshammer">BayesHammer</a> to correct errors in reads.
   After that the corrected reads are stored in the directory <code>&lt;output_dir>/corrected</code> in <code>*.fastq.qz</code> files.
</p>

<p>
    After assembly completion the resulting contigs are stored in the <code>&lt;output_dir>/contigs.fasta</code> file.
    If the <code>--generate-sam-file</code> option was used, the symlink <code>contigs.sam</code> to the SAM-file will also be created in the same directory.
</p>


<h3>2.3 SPAdes command line options </h3>

<p>
    To run SPAdes from the command line type
</p>

<pre class="code">
<code>
    ./spades.py [options] -o &lt;output_dir>
</code>
</pre>

<p>
    To run SPAdes on the toy dataset (see <a href="#sec1.3">section 1.3</a>) type
</p>

<pre class="code">
<code>
    ./spades.py -1 test_dataset/ecoli_1K_1.fq.gz -2 test_dataset/ecoli_1K_2.fq.gz -o spades_test
</code>
</pre>

    <p>Here is the description of options:</p>

<p>
    <code>-o &lt;output_dir> </code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Specify the output directory. Required option.
</p>

<p><code>--sc</code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;This flag is required for MDA (single-cell) data.</p>

<p><code>--12 &lt;filename></code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;File with merged left and right paired end reads.</p>

<p><code>-1  &lt;filename></code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;File with left paired end reads.</p>

<p><code>-2  &lt;filename></code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;File with right paired end reads.</p>

<p><code>-s  &lt;filename></code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;File with unpaired reads.</p>

<p><code>--generate-sam-file</code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Generate a SAM file that contains information about the alignment of the original reads to the resulting
    contigs.</p>

<p><code>-t &lt;int></code> (or <code>--threads &lt;int></code>)<br>
    &nbsp;&nbsp;&nbsp;&nbsp;Number of threads. The default value is 16.</p>

<p><code>-m &lt;int></code> (or <code>--memory &lt;int></code>)<br>
    &nbsp;&nbsp;&nbsp;&nbsp;Sets the memory limit in Gb. SPAdes terminates if it reaches this limit. The default value is 250 Gb.
    Actually consumed physical memory will be smaller than this limit.</p>

<p><code>--tmp-dir &lt;dirname></code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Sets the directory to store temporary files from error correction.  The default value is <code>&lt;output_dir>/corrected/tmp</code>.</p>


<p><code>-k &lt;int,int,...> </code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Comma-separated list of k-mer sizes to be used (all values must be odd,
    less than 100 and listed in the in ascending order). Default is 21,33,55.</p>

<p><code>-i &lt;int></code> (or <code>--iterations &lt;int></code>)<br>
    &nbsp;&nbsp;&nbsp;&nbsp;Number of iterations for error correction. The default value is 1.</p>

<p><code>--phred-offset  &lt;33 or 64> </code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;PHRED quality offset for the input reads, can be either 33 or 64. It will be auto-detected if it is not specified.</p>

<p><code>--only-error-correction</code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Runs error correction only.</p>

<p><code>--only-assembler</code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Runs assembly only.</p>

<p><code>--disable-gzip-output</code> <br>
    &nbsp;&nbsp;&nbsp;&nbsp;Forces error correction not to compress the corrected reads.</p>

<p><code>--test</code> <br>
    &nbsp;&nbsp;&nbsp;&nbsp;Runs SPAdes on the toy dataset, see <a href="#sec1.3">section 1.3</a>.</p>

<p><code>--debug</code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Runs SPAdes in debug mode, keeping intermediate results.</p>

<p><code>-h</code> (or <code>--help</code>)<br>
    &nbsp;&nbsp;&nbsp;&nbsp;Prints help.</p>


<h3>2.4 Files and directories</h3>

<p>
    Usually <code>&lt;output_dir></code> directory contains the following files:
</p>

<pre>
    <code>assembly.log</code> - <i>assembler log</i>
    <code>contigs.fasta</code> - <i>resulting contigs</i>
    <code>contigs.sam</code> - <i>SAM file, generated only with <code>--generate-sam-file</code> option</i>
    <code>corrected/</code> - <i>files from error correction run</i>
        <code>configs/</code> - <i>configuration files for error correction</i>
        <code>correction.log</code> - <i>error correction log</i>
        <code>dataset.info</code> - <i>internal configuration files</i>
        <i>error correction resulting files</i>
    <code>dataset.info</code> - <i>internal configuration files</i>
    <code>K21/</code> - <i>files from the run with K=21</i>
    <code>K33/</code> - <i>files from the run with K=33</i>
    <code>K55/</code> - <i>files from the run with K=55</i>
    <code>params.txt</code> - <i>information about SPAdes parameters in this run</i>
</pre>

<p>
    SPAdes will override these files and directories if they exist in the specified <code>&lt;output_dir></code>.
</p>

<h3>2.5 Assembly evaluation</h3>

<p>
    We recommend to use <a href="http://bioinf.spbau.ru/quast">QUAST</a> for assembly evaluation.
</p>

<h2>3 Feedback and bug reports</h2>

<p>Your comments, bug reports, and suggestions are very welcomed. They will help us to further improve SPAdes.</p>

<p>In case you have troubles running SPAdes, please provide us with the following files from
    <code>&lt;output_dir></code> directory:

<pre class="code">
<code>
    params.txt
    assembly.log
    corrected/correction.log
</code>
</pre>


    <p>Address for communications: <a href="mailto:spades.support@bioinf.spbau.ru">spades.support@bioinf.spbau.ru</a>.</p>

<br/><br/><br/><br/><br/>

</body>
</html>
