<html xmlns="http://www.w3.org/1999/html" xmlns="http://www.w3.org/1999/html" xmlns="http://www.w3.org/1999/html">
<head>
    <title>SPAdes 2.1.0 Manual</title>
    <style type="text/css">
        .code {
            background-color: lightgray;
        }
    </style>
</head>
<body>
<h1>SPAdes 2.1.0 Manual</h1>

<p>SPAdes stands for St. Petersburg genome assembler. It is intended for both single cell MDA and standard
(multicell) assemblies. This manual will help you to install and run SPAdes. You can find the latest version
at <a href="http://bioinf.spbau.ru/spades">http://bioinf.spbau.ru/spades</a>.
</p>


<h2>1. System Requirements</h2>

<p>
    SPAdes requires a 64-bit Linux system. Correcting errors in our E. coli datasets requires about 70 Gb of RAM.
    Assembling our test the <a href="http://spades.bioinf.spbau.ru/spades_test_datasets/E.coli/reads/ECOLI_SC/">single cell  E. coli dataset</a> SPAdes requies 6 Gb of RAM.
    For the <a href="http://spades.bioinf.spbau.ru/spades_test_datasets/E.coli/reads/ECOLI_MC/">multi-cell E. coli dataset</a> it will be about 700 Mb of RAM.
</p>

<h2>2. Installation</h2>

Each of the following subsections describes a way of installing SPAdes. If you have root priveleges then
proceed to either section 2.1 or section 2.2 to install a Debian or RPM package. Section 2.3 provides a way
to build SPAdes from the source code. This does not require root priveleges, but you will have to install also
some libraries that SPAdes depends on. Finally, in section 2.4 it is described how one can just download the
binaries.

<h3>2.1 Installing Debian package</h3>

<p>First, add the repository containing SPAdes by inserting the following line to the end of the file
<code>/etc/apt/sources.list</code>:
</p>

<pre class="code">
<code>
deb http://debian.bioinf.spbau.ru /
</code>
</pre>

<p>Note that the space before the last slash is required. Update the package list by typing</p>

<pre class="code">
<code>
sudo apt-get update
</code>
</pre>

<p>After that, SPAdes can be installed just by typing</p>


<pre class="code">
<code>
sudo apt-get install spades
</code>
</pre>

<h3>2.2 Installing RPM package</h3>

<p>Your can download the RPM package at <a href="http://spades.bioinf.spbau.
ru/release2.1.0/spades-2.1.0.rpm">http://spades.bioinf.spbau.ru/release2.1.0/spades-2.1.0.rpm</a>. The RPM installation procedure varies on different systems, but on
<a href="http://fedoraproject.org/">Fedora</a> we recommend
</p>

<pre class="code">
<code>
sudo yum install spades-2.1.0.rpm
</code>
</pre>

<h3>2.3 Building from source</h3>
<p>If you're unable to install a package (e.g., you don't have root privileges), you can download
<a href="http://spades.bioinf.spbau.ru/release2.1.0/spades_2.1.0.tar.gz">http://spades.bioinf.spbau.ru/release2.1.0/spades_2.1.0.tar.gz</a> and just extract it:
</p>

<pre  class="code">
<code>
wget http://spades.bioinf.spbau.ru/release2.1.0/spades_2.1.0.tar.gz
tar -xzf spades_2.1.0.tar.gz
cd spades_2.1.0
</code>
</pre>

<p>
In this case, you will have to take care of some dependencies beforehand:
<ul>
    <li>g++ (version 4.4 or higher)</li>
    <li>python (version 2.4 or higher)</li>
    <li>cmake (version 2.6 or higher)</li>
    <li>boost (version 1.42 or higher)</li>
    <li>zlib</li>
</ul>
</p>


<h3>2.4 Running binary distribution</h3>

If you don't have possibility to install some dependencies we've prepared the SPAdes
binaries for multiple values of k (see section 5.3), in the range from 11 to 149. For our test datasets, we
recommend k = 21, 33, 55. To use these binaries you need to run the instructions from Section 2.3 and run
the following commands after that:

<pre class="code">
<code>
./spades_download_binary.py 21 33 55
./spades_download_bayeshammer.py
</code>
</pre>

<p>If you need to run SPAdes with other k values you need to download them using <code>./spades_download_binary.py</code> using space-separated values of k as an argument.</p>

<h2>3 Testing your installation</h2>

For testing purposes, SPAdes comes with a toy dataset (first 1000 bp of E. coli). If you installed SPAdes
from package type

<code class="code">
spades.py
</code>

Otherwise, <code>cd</code> to the folder where you extracted <a href="http://spades.bioinf.spbau.ru/release2.1.0/spades_2.1.0.tar.gz">spades_2.1.0.tar.gz</a> to and type

<pre class="code">
<code>
./spades.py
</code>
</pre>

<p>
It runs SPAdes on this toy dataset. If the installation is successful you will see lines like the following lines
in the end of the log:
</p>

<pre class="code">
<code>
 * Corrected reads are in spades_output/ECOLI_1K/corrected/
 * Assembled contigs are spades_output/ECOLI_1K/spades_04.18_17.59.30/ECOLI_1K.fasta
 * Assessment of their quality is spades_output/ECOLI_1K/spades_04.18_17.59.30/quality_results/quality.txt

Thank you for using SPAdes!

======= SPAdes pipeline finished
</code>
</pre>

<h2>4 Running SPAdes</h2>

<p>
To run SPAdes type
</p>

<pre class="code">
<code>
spades.py &lt;config.info file>
</code>
</pre>

<p>
if you installed SPAdes from package (see sections 2.1 and 2.2), or
</p>

<pre class="code">
<code>
./spades.py &lt;config.info file>
</code>
</pre>

<p>
from the directory you extracted the archive (see sections 2.3 and 2.4). By default (i.e., if no config file is given)
SPAdes uses the file <code>spades_config.info</code>. If SPAdes was installed from package the default config file is
located at <code>/usr/share/spades/spades_config.info</code>, otherwise it can be found in the directory where SPAdes
was extracted to. Below we first give an example of a config file and then explain its contents in detail. In config files,
the part of a line starting with a semicolon is a comment.
</p>

<pre class="code">
<code>
; all paths are relative to the working directory but you can also use $cfg variable
; to set paths relative to this config
project_name ECOLI_1K ; optional. Name of this config file is used if not specified
output_dir ./spades_output

dataset
{
    paired_reads $cfg/test_dataset/ecoli_1K_1.fq.gz $cfg/test_dataset/ecoli_1K_2.fq.gz
    single_reads ; optional
    single_cell false
}

error_correction
{
    tmp_dir ; optional. &lt;output_dir>/corrected/tmp is used if not specified
    qvoffset ; optional. Either 33 or 64
    max_iterations 2
    max_threads 16
    max_memory 250 ; in GB
    gzip_output true ; optional
}

assembly
{
    iterative_K 21 33 55
    generate_sam_files true ; optional. Default is false
    gap_closer true ; optional. Default is true
}

; If this section exists (even if it is empty), spades.py
; evaluates quality of the final contigs
quality_assessment
{
    reference $cfg/test_dataset/reference_1K.fa.gz ; optional
    genes $cfg/test_dataset/genes_1K.txt ; optional
    operons $cfg/test_dataset/operons_1K.txt ; optional
}
</code>
</pre>

<p>
There are two global variables: <code>project_name</code> specifies the name of the project and <code>output_dir</code>
specifies the output directory. It is recommended to set a meaningful name for each
separate project.
</p>

<p>
The first section in the config file is the required <code>dataset</code> section that describes the dataset.
The other sections <code>error_correction</code>, <code>assembly</code>, and <code>quality_assessment</code> correspond
to three stages of SPAdes pipeline. To run only some of these three stages just remove (or
comment or rename) unnecessary stages. If you run SPAdes with the same output_dir and project_name
as one of previous runs, it checks if you have already made the error correction stage and asks if you want
to correct the reads again.
</p>

<h3>4.1 Dataset settings</h3>
<p>
SPAdes accepts single reads as well as forward-reverse paired end reads in FASTA and FASTQ format;
however, in order to run error correction, reads should be in FASTQ format. All files may be compressed
with gzip. At present, SPAdes can accept only one paired-end library as input.
</p>

<p>
<code>paired_reads</code> specifies input paired end reads organized into one or two files.
    <ul>
        <li>If using a single file, interlace the corresponding mated "1" and "2" reads. Put the read from the Illumina
            file <code>s_n_1_export.txt</code> first, followed by the read from the Illumina file <code>s_n_2_export.txt</code>.</li>
        <li>If using a pair of files, the first file should have the Illumina "1" subreads and the second file
            should have the Illumina "2" subreads, with corresponding reads in the exact same order. The
            two filenames should be separated by a space.</li>
    </ul>
</p>

<p>
<code>single_reads</code> specifies reads without pairs. You can specify any number of files with single reads. File
names should be separated by spaces.
</p>

<p>
<code>single_cell</code> flag is set if input data was obtained with MDA (single cell) technology.
</p>

<h3>4.2 Error correction settings</h3>
<p>
The <code>error_correction</code> section is responsible for correcting errors in the input reads; error correction in the
SPAdes pipeline is done with <a href="http://bioinf.spbau.ru/en/spades/bayeshammer">BayesHammer</a>,
a recently developed error correction tool that works well for the case of non-uniform coverage (e.g., on single-cell data). Note that for an accurate assembly it is important
to use this stage, especially for single-cell datasets.
</p>

<p>
<code>tmp_dir</code> specifies the directory where temporary files are stored. Note that BayesHammer uses a significant
    amount of disk space (even though it cleans up after itself).
</p>

<p>
    <code>qvoffset</code> sets PHRED quality offset in the input reads (33 or 64). If not given, BayesHammer tries to find
    this value automatically.
</p>

<p>
    <code>max_iterations</code> sets the maximum number of iterations for the error correction procedure. Usually, the
    second iteration does improve upon the first, but then there is almost no change.
</p>

<p>
    <code>max_threads</code> sets the maximum number of threads.
</p>

<p>
<code>max_memory</code> sets the maximum amount of available RAM (in Gb). BayesHammer (and the entire SPAdes
    pipeline) will fail if it attempts to allocate more RAM than specified here.
</p>

<p>
    <code>gzip_output</code> flag forces BayesHammer to output corrected reads compressed with gzip.
</p>

<h3>4.3 Assembly settings</h3>

<p>
    The <code>assembly</code> section controls the assembly process.
</p>

<p>
    <code>iterative_K</code> specifies one or more odd values for k-mer (vertex) sizes. Informally, smaller values of k make
    the graph more connected, but at the same time more tangled, while higher values of k may defragment
    the graph, but allow to resolve short repeats. The typical value for this variable is <code>21 33 55</code>. See the
    paper [1] for more details. Note that in the configuration file, k is the size of a vertex, while in the
    paper, k is the size of an edge (i.e., larger by one).
</p>

<p>
<code>paired_mode</code> turns the repeat resolver on/off if parameter set to true/false.
</p>

<p>
    <code>generate_sam_files</code> flag forces SPAdes to generate a SAM-file showing how the paired reads specified in
    dataset section of config file are aligned to resulting contigs. Alignments from this SAM-file can be
    visualized, e.g., by Tablet [3].
</p>

<p>
    gap_closer turns the gap closer on/off. This module restores k-mers that are not covered not by whole
    reads but by overlapping parts of two reads and are supported by paired reads. This is efficient when
    a library has a small insert size (up to 3x read length). If insert size is large, you can turn it off to
    slightly decrease the runing time of SPAdes. Note: you can decrease the running time of SPAdes by
    setting just one value of k (say, 55) and turning the gap closer on (since the gap closer usually gives
    similar results as running with several smaller values of k).
</p>



<h3>4.4 Quality assessment settings</h3>
<p>
    The <code>quality_assessment</code> section is needed to estimate the quality of the resulting assembly.
    If this section is present in the config file then SPAdes evaluates assembly results by computing various
    metrics (e.g., N50, NG50, number of long ORFs, etc.) on the resulting contigs.
</p>

<p>
    <code>reference</code>,  <code>genes</code> and <code>operons</code> variables specify files with
    reference, genes and operons.
</p>
<p>
    <code>genes</code> and <code>operons</code> should be either in GFF format <a href="http://www.sanger.ac.uk/resources/
    software/gff/spec.html">version 2</a>  or <a href="http://www.sequenceontology.org/gff3.shtml">version 3</a> or in
    plain text format with 4 tab-separated columns: <strong>the chromosome name</strong>(it should match the exact string
    in the reference file), <strong>gene or operon id</strong>, <strong>start position</strong> and <strong>end postion</strong> (the coordinates are 1-based
    and if start postion less than end postion than that gene or operon is on the positive strand and it is on the
    negative strand otherwise).
</p>

<p>
    Note that the quality is estimated even if this section is empty. Our quality tool uses <a href="http://www.plantagora.org/">Plantagora</a>
    and <a href="http://mummer.sourceforge.net/">MUMmer</a>. The computed metrics are described in detail in section 5.
</p>

<h3>4.5 Creating a config file for your dataset</h3>
<p>
    To create a config file for your dataset <code>cd</code> to the SPAdes directory and copy the default config file to a new file:
</p>

<pre class="code">
<code>
    cp spades_config.info &lt;your_project_name.info file>
</code>
</pre>

<p>
    Then edit this new file according to your dataset.
</p>

<h2>5 Assembly quality metrics</h2>

<p>
    In this section, we describe various metrics computed by the quality tool of SPAdes pipeline (see section 4.4).
    Note that the metrics based on the reference are computed only if the reference is given in the config file.
    N50 is the contig length such that using longer or equal length contigs produces half (50%) the bases of
    the assembly. Usually there is no value that produces exactly 50%, so the more technical definition is
    the minimal length x such that using contigs of length at least x accounts for at least 50% of the total
    assembly length.
</p>

<p>
    <strong>NG50</strong> is the contig length such that using longer or equal length contigs produces half (50%) the bases of
    the reference genome.
</p>

<p>
    <strong>N75</strong> and </strong>NG75</strong> are defined similarly with 75% instead of 50%.
</p>

<p>
    <strong>Number of contigs</strong> is the total number of contigs in the assembly.
</p>

<p>
    <strong>Largest contig</strong> is the length of the longest contig in the assembly.
</p>

<p>
    <strong>Total length</strong> is the total number of bases in the assembly.
</p>

<p>
    <strong>Reference length</strong> is the total number of bases in the reference.
</p>

<p>
    <strong>Average %IDY</strong> is the average of alignment identity percent (quality) among all contigs.
</p>

<p>
    <strong>Misassemblies</strong> gives the number of positions at the assembled contigs where the left flanking sequence
    aligns over 1kb away from the right flanking sequence on the reference or they overlap on more than
    1 kb or flanking sequences align on different strands.
</p>

<p>
    <strong>Misassembled contigs</strong> is the number of contigs that contain misassembly events.
</p>

<p>
    <strong>Misassembled contig bases</strong> is the number of total bases contained in all contigs that have one or more
    misassemblies.
</p>

<p>
    <strong>Unaligned contigs</strong> is the number of contigs that have no alignment to the reference sequence. The value
    "X (Y partial) (Z)" means: X totally unaligned contigs plus Y partially unaligned contigs and the
    total number of unaligned bases is Z.
</p>

<p>
    <strong>Ambiguous contigs</strong> is the number of contigs that have reference alignments of equal quality in multiple
    locations on the reference. The value "X (Y)" means X ambiguous contigs with Y bases in them.
</p>

<p>
    <strong>NA50 (A stands for "aligned")</strong> is defined similarly to N50 but only lengths of aligned blocks are counted.
    If a contig has a misassembly with respect to the reference, it is broken into smaller pieces. If a contig
    has bases that do not align to the reference, they are not counted in NA*, NGA*. NGA50, NA75, NGA75 are defined in the same way.
</p>

<p>
    <strong>Mapped genome (%)</strong> is the ratio of the total number of aligned bp in the assembly to the genome size.
    Short contigs (less than 500 bp) that map to multiple places may be counted many times in this
    quantity. A base in the genome is counted as aligned one if there is at least one contig with at least
    one alignment with this base.
</p>

<p>
    <strong>Genes</strong> is the number of genes in the assembly (full and partial), based on the positions the contigs map
    to in the reference genome with an annotated list of genes. The value "X + Y partial" means X full
    genes and Y partial ones (at least 100 bases are aligned).
</p>

<p>
    <strong>Operons</strong> is the number of operons in the assembly (full and partial), based on the positions the contigs
    map to in the reference genome with an annotated list of operons. The value "X + Y partial" means
    X full operons and Y partial ones (at least 100 bases are aligned).
</p>

<p>
    <strong>Number of miscalled</strong> is the number of miscalled bases (e.g. having "C" in the assembly and "G" in the
    genome).
</p>

<p>
    <strong>Number of uncalled</strong> is the number of uncalled bases ("N").
    Bases missed is the number of missed bases (which are presented in the genome and missed in the assembly).
</p>

<h2>6 Output</h2>
    Results can be found in the directory specified by <code>output_dir</code>. Specific directories containing corrected reads,
    assembly, and quality estimate are shown at the end of the log.

<h2>7 Postprocessing</h2>

    <h3>7.1 Correcting errors in the resulting contigs</h3>
    While SPAdes produces accurate assemblies, we recommend running <a href="http://bix.ucsd.edu/SEQuel/">SEQuel</a> [4]
    after SPAdes to further  reduce the number of small errors (single nucleotide substitutions and small indels).
    Note that one can set the flag generate_sam to SPAdes (see section 4.3)
    to generate a SAM-file showing how the reads are aligned to the resulting contigs. This file can then be
    given to SEQuel as input.

    <h3>7.2 Scaffolding</h3>
    The current version of SPAdes does not have a scaffolding stage. One can use a separate scaffolder such as
    <a href="http://sourceforge.net/projects/operasf/">Opera</a> [2].

<h2>8 Feedback and bug reports</h2>
    We will be thankful if you help us make SPAdes better by sending your comments, bug reports, and suggestions
    to <a href="mailto:spades.support@bioinf.spbau.ru">spades.support@bioinf.spbau.ru</a>.


<h2>References</h2>
<p>
    [1] A. Bankevich, S. Nurk, et al. SPAdes: a New Genome Assembly Algorithm and its Applications to
    Single-Cell Sequencing. Journal of Computational Biology, 2012. To appear.
</p>
<p>
    [2] S. Gao, WK. Sung, and N. Nagarajan. Opera: reconstructing optimal genomic scaffolds with highthroughput
    paired-end sequences. Journal of Computational Biology, 18(11):1681–1691, 2011.
</p>
<p>
    [3] I. Milne, M. Bayer, L. Cardle, P. Shaw, G. Stephen, F.Wright, and D. Marshall. Tablet—next generation
    sequence assembly visualization. Bioinformatics, 26(3):401–402, 2010.
</p>
<p>
    [4] R. Ronen, C. Boucher, H. Chitsaz, and P. Pevzner. SEQuel: Improving the accuracy of genome assemblies.
    Bioinformatics, 2012. To appear.
</p>



</body>
</html>
