<html>
<head>
    <title>SPAdes 2.2.0 Manual</title>
    <style type="text/css">
        .code {
            background-color: lightgray;
        }
    </style>
</head>
<body>
<h1>SPAdes 2.2.0 Manual</h1>

<p>SPAdes stands for St. Petersburg genome assembler. It is intended for both standard (multicell) and single-cell MDA
    bacteria  assemblies. This manual will help you to install and run SPAdes. You can find the latest SPAdes release
at <a href="http://bioinf.spbau.ru/spades">http://bioinf.spbau.ru/spades</a>. The latest version of this manual can be found <a href="http://spades.bioinf.spbau.ru/release2.2.0/manual.html">here</a>.
    SPAdes version 2.2.0 released under  GPLv2 on 2 August 2012.
</p>


<h2>1. Installation</h2>

<p>
    SPAdes requires a 64-bit Linux system.
    We have two test datasets: <a href="http://spades.bioinf.spbau.ru/spades_test_datasets/ecoli_sc/">single-cell  E. coli dataset</a> and
    <a href="http://spades.bioinf.spbau.ru/spades_test_datasets/ecoli_mc/">multicell E. coli dataset</a>. SPAdes requires 35 Gb of RAM for processing these datasets. Also SPAdes requires  Python 2.4 or higher installed.
</p>

<h3>1.1 SPAdes tar.gz file</h3>

<p>To download <a href="http://spades.bioinf.spbau.ru/release2.2.0/spades-2.2.0.tar.gz">SPAdes tar.gz</a> and extract it:
</p>

<pre  class="code">
<code>
    wget http://spades.bioinf.spbau.ru/release2.2.0/spades-2.2.0.tar.gz
    tar -xzf spades-2.2.0.tar.gz
    cd spades-2.2.0
</code>
</pre>

<h3>1.2 Getting SPAdes binaries</h3>

    <p>There are two ways to obtain SPAdes binaries: download static builds from our server or compile SPAdes on you server.</p>

    <p>We recommend to download binaries with the following script:</p>

<pre class="code">
<code>
    ./spades_download_binary.py
</code>
</pre>

<p>
    Also you can compile SPAdes yourself, but SPAdes depends on the following libraries for compiling its code:
</p>

    <ul>
        <li>g++ (version 4.4 or higher)</li>
        <li>cmake (version 2.6 or higher)</li>
        <li>boost (version 1.42 or higher)</li>
        <li>zlib</li>
    </ul>

<p>
    If you meet this requirements you can build SPAdes with the following script:
</p>

<pre class="code">
<code>
    ./spades_compile.sh
</code>
</pre>

<p>In both cases you should get <code>bin</code> directory with files <code>hammer</code> (error correcting module) ans <code>spades</code> (assembly module).
</p>

<a name="sec1.3"></a>
<h3>1.3 Testing your installation</h3>

<p>
For testing purposes, SPAdes comes with a toy dataset (first 1000 bp of E. coli). If you run <code>spades.py</code> with the parameter <code>--test</code>
</p>

<pre class="code">
<code>
    ./spades.py --test
</code>
</pre>

<p>
it will process this dataset and if the installation is successful you will see something like this at the end of the log:
</p>

<pre class="code">
<code>
 * Corrected reads are in spades_output/ECOLI_1K/corrected/
 * Assembled contigs are spades_output/ECOLI_1K/spades_04.18_17.59.30/ECOLI_1K.fasta

Thank you for using SPAdes!

======= SPAdes pipeline finished
</code>
</pre>

<h2>2 Running SPAdes</h2>

<h3>2.1 Input data</h3>
<p>
    SPAdes accepts single reads as well as forward-reverse paired end reads in FASTA and FASTQ format;
    however, in order to run error correction, reads should be in FASTQ format. All files may be compressed
    with gzip. At present, SPAdes can accept only one paired-end library as an input.
</p>

<p>
    SPAdes supports paired end reads organized in two separate files or combined in one:

<ul>

    <li>If the forward and reverse reads for each lane are split into separate FASTQ files (e.g.
        <code>s_1_1_sequence.txt</code> and <code>s_1_2_sequence.txt</code> for the forward and reverse reads, respectively, from lane 1),
        the second file  should have corresponding reads in the exact same order.
    </li>
    <li>
        If using a single file containing reads in both orientations, interlace the corresponding reads such that each read from
        <code>s_1_1_sequence.txt</code> is followed by the corresponding read from <code>s_1_2_sequence.txt</code>.
    </li>


</ul>

<h3>2.2 SPAdes pipeline</h3>

<p>SPAdes automatically creates a <code>spades_output</code> directory where it stores all of the resulting files. For each dataset SPAdes creates
    a separate directory <code>&lt;project_name></code> in <code>spades_output</code>.
</p>

<p>Before starting assembly SPAdes initiates <a href="http://bioinf.spbau.ru/en/spades/bayeshammer">BayesHammer</a> to correct errors in reads.
   After that the corrected reads are stored in the directory <code>spades_output/project_name/corrected</code> in <code>*.fastq.qz</code> files.
</p>

<p>
    After assembly completion, the resulting contigs are stored in the directory <code>spades_output/project_name/contigs</code>  in <code>&lt;project_name>.fasta</code> file.
    If the   <code>--generate-sam-file</code> option was used, the symlink to the <code>&lt;project_name>.sam</code> will also be created in the same directory.
</p>

<p>
    If user runs SPAdes with the same <code>project_name</code> more than once, SPAdes finds the corrected reads and asks the user whether
    it should start error correction again.   By default, we do not re-run error correction and we start the assembly in 10 seconds
    unless user says otherwise.
</p>

<p>
    There are two ways to run SPAdes for you dataset: you can specify the parameters from the command line or provide the configuration
    file as the only <code>./spades.py</code> parameter.
</p>



<h3>2.3 SPAdes command line options </h3>

<p>
    To run SPAdes from the command line type
</p>

<pre class="code">
<code>
    ./spades.py [options] -n &lt;project_name>
</code>
</pre>

<p>
    To run SPAdes on the toy dataset (see <a href="#sec3">section 3</a>) type
</p>

<pre class="code">
<code>
    ./spades.py -1 test_dataset/ecoli_1K_1.fq.gz -2 test_dataset/ecoli_1K_2.fq.gz -n ECOLI_1K
</code>
</pre>

    <p>Here is the description of options:</p>

<p>
    <code>-n &lt;project_name></code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;A required option that sets the name of the project.</p>

<p>
    <code>-o &lt;output_dir> </code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Specify the output directory. Default: <code>spades_output</code>.
</p>

<p><code>--sc</code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;This flag is required for MDA (single-cell) data.</p>

<p><code>--12 &lt;filename></code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;File with merged left and right paired end reads.</p>

<p><code>-1  &lt;filename></code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;File with left paired end reads.</p>

<p><code>-2  &lt;filename></code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;File with right paired end reads.</p>

<p><code>-s  &lt;filename></code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;File with unpaired reads.</p>

<p><code>--generate-sam-file</code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Generate a SAM file. See more details in <a href="#sec3">section 3</a>.</p>

<p><code>-t &lt;int></code> (or <code>--threads &lt;int></code>)<br>
    &nbsp;&nbsp;&nbsp;&nbsp;Number of threads. The default value is 16.</p>

<p><code>-m &lt;int></code> (or <code>--memory &lt;int></code>)<br>
    &nbsp;&nbsp;&nbsp;&nbsp;Sets the memory limit in Gb. SPAdes terminates if it reaches this limit. The default value is 250 Gb.
    Actually consumed physical memory will be smaller than this limit.</p>

<p><code>--tmp-dir &lt;dirname></code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Sets the directory to store temporary files from error correction.  The default value is <code>&lt;output_dir>/&lt;project_name>/corrected/tmp</code>.</p>


<p><code>-k &lt;int,int,...> </code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Comma-separated list of k-mer sizes to be used (all values must be odd,
    less than 100 and listed in the in ascending order). Default is 21,33,55.</p>

<p><code>-i &lt;int></code> (or <code>--iterations &lt;int></code>)<br>
    &nbsp;&nbsp;&nbsp;&nbsp;Number of iterations for error correction. The default value is 1.</p>

<p><code>--phred-offset  &lt;33 or 64> </code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;PHRED quality offset for the input reads, can be either 33 or 64. It will be auto-detected if it is not specified.</p>

<p><code>--only-error-correction</code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Runs error correction only.</p>

<p><code>--only-assembler</code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Runs assembly only.</p>

<p><code>--disable-gzip-output</code> <br>
    &nbsp;&nbsp;&nbsp;&nbsp;Forces error correction not to compress the corrected reads.</p>

<p><code>--test</code> <br>
    &nbsp;&nbsp;&nbsp;&nbsp;Runs SPAdes on the toy dataset, see <a href="#sec1.3">section 1.3</a> .</p>

<p><code>--debug</code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Runs SPAdes in debug mode, keeping intermediate results.</p>

<p><code>-h</code> (or <code>--help</code>)<br>
    &nbsp;&nbsp;&nbsp;&nbsp;Prints help.</p>


<h3>2.4 Preparing configuration files</h3>

<p>
 You can find the toy dataset configuration file <code>spades_config.info</code> (<a href="#sec1.3">section 1.3</a>)
    in the directory containing extracted archive (see section 1.1). This file can be used as the template for your datasets.
    Text placed after the first semicolon in this file is a comment.
</p>

<p>
    The configuration file starts with the common parameters for the SPAdes run. We recommend setting the <code>project_name</code> parameter that specifies the directory
    and names for output files. Next is a dataset section that contains input reads information. This section is required and you need to specify
    either <code>paired_reads</code> or <code>single_reads</code> or both.
</p>

<p>
    The next two sections are <code>error_correction</code> and <code>assembly</code>. If you want to skip one of these stages, you can
    remove or rename or comment it.
</p>



<h3>2.5 Supporting files and directories</h3>

<p>
    For each run SPAdes creates the directory <code>spades_&lt;date_time></code> in <code>spades_output/&lt;project_name></code>.
    It contains internal configs,  logs and intermediate results for different values of k.
    Also symlinks <code>latest</code> and <code>latest_success</code> are created for the the directories
    for the latest and the latest successful runs.
</p>


<a name="sec3"></a>
<h2>3 Postprocessing</h2>

<p>
    After running SPAdes you can evaluate different quality metrics. Please see <a href="quality.html">quality.html</a> for details.
</p>

<p>
    SPAdes generates a SAM file for further processing if the <code>--generate-sam-file</code> option in set the command line.
    SAM file contains information about the alignment of the original reads to the resulting
    contigs.  We recommend using <a href="http://bioinf.scri.ac.uk/tablet/">Tablet</a> [3] to visualize  these alignments.
</p>

<p>
    Also we recommend to use  <a href="http://bix.ucsd.edu/SEQuel/">SEQuel</a> [4]
    after SPAdes to further  reduce the number of small errors (single nucleotide substitutions and small indels). It also requires the SAM-file
    as an input.
</p>
<p>
    Publicly available tools (for example, <a href="http://sourceforge.net/projects/operasf/">Opera</a> [2])
    can be used to build scaffolds as the current version of SPAdes is not yet providing this option.
</p>

<h2>4 Feedback and bug reports</h2>

<p>Your comments, bug reports, and suggestions are very welcomed. They will help us to further improve SPAdes.</p>

<p>In case you have troubles running SPAdes, please provide us with the following files from
    <code>&lt;output_dir>/&lt;project_name></code> directory:

<pre class="code">
<code>
    params.txt
    corrected/correction.log
    contigs/assembly.log
</code>
</pre>


    <p>Address for communications: <a href="mailto:spades.support@bioinf.spbau.ru">spades.support@bioinf.spbau.ru</a>.</p>

<h2>References</h2>
<p>
    [1] A. Bankevich, S. Nurk, et al. <a href="http://online.liebertpub.com/doi/abs/10.1089/cmb.2012.0021">SPAdes: A New Genome Assembly Algorithm and Its Applications to Single-Cell Sequencing.</a>
    Journal of Computational Biology. May 2012, 19(5): 455-477. doi:10.1089/cmb.2012.0021.
</p>
<p>
    [2] Song Gao, Wing-Kin Sung, and Niranjan Nagarajan. <a href="http://online.liebertpub.com/doi/abs/10.1089/cmb.2011.0170">Opera: reconstructing optimal genomic scaffolds with high-throughput paired-end sequences.</a>
    Journal of Computational Biology. November 2011, 18(11): 1681-1691. doi:10.1089/cmb.2011.0170.
</p>
<p>
    [3] I. Milne, M. Bayer, L. Cardle, P. Shaw, G. Stephen, F.Wright, and D. Marshall. <a href="http://bioinformatics.oxfordjournals.org/content/26/3/401.abstract">Tablet &mdash; next generation
    sequence assembly visualization.</a> Bioinformatics (2010) 26 (3): 401-402.   doi: 10.1093/bioinformatics/btp666
</p>
<p>
    [4] R. Ronen, C. Boucher, H. Chitsaz, and P. Pevzner. SEQuel: <a href="http://bioinformatics.oxfordjournals.org/content/28/12/i188.abstract">Improving the accuracy of genome assemblies.</a>
    Bioinformatics (2012) 28 (12): i188-i196.   doi: 10.1093/bioinformatics/bts219
</p>

<br/><br/><br/><br/><br/>

</body>
</html>
