<html>
<head>
    <title>SPAdes 2.1.0 Manual</title>
    <style type="text/css">
        .code {
            background-color: lightgray;
        }
    </style>
</head>
<body>
<h1>SPAdes 2.1.0 Manual</h1>

<p>SPAdes stands for St. Petersburg genome assembler. It is intended for both standard (multicell) and single-cell MDA
    bacteria  assemblies. This manual will help you to install and run SPAdes. You can find the latest SPAdes release
at <a href="http://bioinf.spbau.ru/spades">http://bioinf.spbau.ru/spades</a>. The latest version of this manual can be found <a href="http://spades.bioinf.spbau.ru/release2.1.0/manual.html">here</a>.
</p>


<h2>1. Installation</h2>

<p>
    SPAdes requires a 64-bit Linux system. Correcting errors in our E. coli datasets requires about 70 Gb of RAM.
    Assembling our test <a href="http://spades.bioinf.spbau.ru/spades_test_datasets/E.coli/reads/ECOLI_SC/">single-cell  E. coli dataset</a> with SPAdes requires 6 Gb of RAM.
    For the <a href="http://spades.bioinf.spbau.ru/spades_test_datasets/E.coli/reads/ECOLI_MC/">multicell E. coli dataset</a> it will be about 700 Mb of RAM. Also SPAdes requires
    Python 2.4 or higher installed.
</p>

<h3>1.1 SPAdes tar.gz file</h3>

<p>You can download <a href="http://spades.bioinf.spbau.ru/release2.1.0/spades_2.1.0.tar.gz">SPAdes tar.gz</a> and extract it:
</p>

<pre  class="code">
<code>
    wget http://spades.bioinf.spbau.ru/release2.1.0/spades_2.1.0.tar.gz
    tar -xzf spades_2.1.0.tar.gz
    cd spades_2.1.0
</code>
</pre>

<p>
    SPAdes depends on the following libraries for compiling its code:
</p>

    <ul>
        <li>g++ (version 4.4 or higher)</li>
        <li>cmake (version 2.6 or higher)</li>
        <li>boost (version 1.42 or higher)</li>
        <li>zlib</li>
    </ul>

<p>If you are not able to match these requirements please use downloaded binaries as described in the next section.</p>

<h3>1.2 Downloading binaries</h3>

<p>If you want to run SPAdes with default parameters you download the required binaries with the following scripts: </p>

<pre class="code">
<code>
    ./spades_download_binary.py 21 33 55
    ./spades_download_bayeshammer.py
</code>
</pre>

<p>It will download binaries to <code>.spades/release2.1.0</code> directory in your user home.
SPAdes uses binaries optimized for each k value and you can download binaries
for the required k values with <code>./spades_download_binary.py</code> script using space-separated values of k as an argument.</p>


<h3>1.3 Testing your installation</h3>

<p>
For testing purposes, SPAdes comes with a toy dataset (first 1000 bp of E. coli). If you run <code>spades.py</code> with parameter <code>--test</code>
</p>

<pre class="code">
<code>
    ./spades.py --test
</code>
</pre>

<p>
it will process this dataset
and if the installation is successful you will see something like this at the end of the log:
</p>

<pre class="code">
<code>
 * Corrected reads are in spades_output/ECOLI_1K/corrected/
 * Assembled contigs are spades_output/ECOLI_1K/spades_04.18_17.59.30/ECOLI_1K.fasta

Thank you for using SPAdes!

======= SPAdes pipeline finished
</code>
</pre>

<h2>2 Running SPAdes</h2>

<h3>2.1 Input data</h3>
<p>
    SPAdes accepts single reads as well as forward-reverse paired end reads in FASTA and FASTQ format;
    however, in order to run error correction, reads should be in FASTQ format. All files may be compressed
    with gzip. At present, SPAdes can accept only one paired-end library as input.
</p>

<p>
    SPAdes supports paired end reads organized into one or two files.

<ul>
    <li>If using a single file, interlace the corresponding mated "1" and "2" reads. Put the read from the Illumina
        file <code>s_n_1_export.txt</code> first, followed by the read from the Illumina file <code>s_n_2_export.txt</code>.</li>
    <li>If using a pair of files, the first file should have the Illumina "1" subreads and the second file
        should have the Illumina "2" subreads, with corresponding reads in the exact same order. </li>
</ul>

<h3>2.2 SPAdes pipeline</h3>
<p>
    SPAdes has two stages: error correction by <a href="http://bioinf.spbau.ru/en/spades/bayeshammer">BayesHammer</a> and genome assembly. By default all the
    results are stored in <code>spades_output</code> directory. For each dataset SPAdes creates separate directory <code>&lt;project_name></code> in <code>spades_output</code>.
    Corrected reads are stored in <code>corrected</code> subdirectory in <code>*.fastq.qz</code> files.
    The resulting contigs are stored in <code>contigs</code> subdirectory in <code>*.fasta</code> file.
</p>

<p>
    If SPAdes finds the corrected reads it asks user whether it should start error correction again. By default we do not re-run error correction and we start the assembly in 10 seconds
    unless user says otherwise.
</p>

<p>
    There are two ways to run SPAdes for you dataset you can specify the parameters from the command line or provide the configuration file with <code>.info</code> extension
    as the only <code>./spades.py</code> parameter.
</p>



<h3>2.3 SPAdes command line options </h3>
<p>


</p>



<h3>2.4 Preparing configuration files</h3>

<p>
You can find the configuration file <code>spades_config.info</code> for the toy dataset from section 1.3 in the directory you extracted the archive (see section 1.1).
You can use this file as the template for your datasets. In  this file in each line all text after the first semicolon is a comment.
</p>

<p>
    The configuration file starts with the common parameters for the SPAdes run. We do recommend to set the <code>project_name</code> parameter that specifies the directory
    and names for output files. After that there is a dataset section that contains the information about input reads. This section is required and you need to specify
    either <code>paired_reads</code> or <code>single_reads</code> or both
</p>

<p>
    The next two sections are <code>error_correction</code> and <code>assembly</code>. If you want to skip one of these stages you can
    remove or rename or comment it.
</p>



<h3>2.5 Supporting files and directories</h3>

<p>
    SPAdes uses <code>~/.spades</code> directory to store compiled binaries and reuse them. If you want to save disk space or just want to be sure that
    you use recently compiled binaries you can just delete this directory. But we suppose that you don't care about it.
</p>

<p>
    For each SPAdes run we create directory <code>spades_&lt;date_time></code> in <code>spades_output/&lt;project_name></code>. It contains internal configs,
    logs and intermediate results for different k values. Also we have symbolic links <code>latest</code> and <code>latest_success</code> to the directories
    for latest and latest successful runs.
</p>



<h2>3 Postprocessing</h2>
<p>
    While SPAdes produces accurate assemblies, we recommend running <a href="http://bix.ucsd.edu/SEQuel/">SEQuel</a> [4]
    after SPAdes to further  reduce the number of small errors (single nucleotide substitutions and small indels).
    Note that one can set the <code>--generate_sam_file</code> to SPAdes (see section 2.3)
    to generate a SAM-file showing how the reads are aligned to the resulting contigs. This file can then be
    given to SEQuel as input.
</p>
<p>
    The current version of SPAdes does not have a scaffolding stage. One can use a separate scaffolder such as
    <a href="http://sourceforge.net/projects/operasf/">Opera</a> [2].
</p>

<h2>5 Feedback and bug reports</h2>
<p>
    After running SPAdes you can evaluate different quality metrics using our pipeline. Please see <a href="quality.html">quality.html</a> for details.
</p>

<p>
    We will be thankful if you help us make SPAdes better by sending your comments, bug reports, and suggestions
    to <a href="mailto:spades.support@bioinf.spbau.ru">spades.support@bioinf.spbau.ru</a>.
</p>

<p>
    We kindly ask you to attach logs <code>corrected/correction.log</code> and <code>contigs/assembly.log</code> if you have troubles running SPAdes.
</p>


<h2>References</h2>
<p>
    [1] A. Bankevich, S. Nurk, et al. <a href="http://online.liebertpub.com/doi/abs/10.1089/cmb.2012.0021">SPAdes: A New Genome Assembly Algorithm and Its Applications to Single-Cell Sequencing.</a>
    Journal of Computational Biology. May 2012, 19(5): 455-477. doi:10.1089/cmb.2012.0021.
</p>
<p>
    [2] Song Gao, Wing-Kin Sung, and Niranjan Nagarajan. <a href="http://online.liebertpub.com/doi/abs/10.1089/cmb.2011.0170">Opera: reconstructing optimal genomic scaffolds with high-throughput paired-end sequences.</a>
    Journal of Computational Biology. November 2011, 18(11): 1681-1691. doi:10.1089/cmb.2011.0170.
</p>
<p>
    [3] I. Milne, M. Bayer, L. Cardle, P. Shaw, G. Stephen, F.Wright, and D. Marshall. <a href="http://bioinformatics.oxfordjournals.org/content/26/3/401.abstract">Tabletâ€”next generation
    sequence assembly visualization.</a> Bioinformatics (2010) 26 (3): 401-402.   doi: 10.1093/bioinformatics/btp666
</p>
<p>
    [4] R. Ronen, C. Boucher, H. Chitsaz, and P. Pevzner. SEQuel: Improving the accuracy of genome assemblies.
    Bioinformatics, 2012. To appear.
</p>



</body>
</html>
