<html>
<head>
    <title>truSPAdes 1.0 Manual</title>
    <style type="text/css">
        .code {
            background-color: lightgray;
        }
    </style>
</head>
<body>
<h1>SPAdes 1.0 Manual</h1>

1. <a href="#sec1">About truSPAdes</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;1.1. <a href="#sec1.1">TruSeq data</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;1.2. <a href="#sec1.2">TruSPAdes performance</a><br>
2. <a href="#sec2">Installation</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;2.1. <a href="#sec2.1">Downloading truSPAdes Linux binaries</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;2.2. <a href="#sec2.2">Downloading truSPAdes binaries for Mac</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;2.3. <a href="#sec2.3">Downloading and compiling truSPAdes source code</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;2.4. <a href="#sec2.4">Verifying your installation</a><br>
3. <a href="#sec3">Running truSPAdes</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;3.1. <a href="#sec3.2">TruSPAdes command line options</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;3.2. <a href="#sec3.5">TruSPAdes output</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;3.3. <a href="#sec3.6">Assembly evaluation</a><br>
4. <a href="#sec5">Feedback and bug reports</a><br>
<br>

<a name="sec1"></a>
<h2>1. About truSPAdes</h2>
<p>
    TruSPAdes is an assembler for short reads produced by Illumina TruSeq technology. TruSPAdes accepts as an input reads from single TruSeq barcodes and assembles long virtual reads. This manual will help you to install and run truSPAdes. TruSPAdes version 1.0 was released under GPLv2 on January 16, 2015 and can be downloaded from  <a href="http://bioinf.spbau.ru/en/truspades" target="_blank">http://bioinf.spbau.ru/en/truspades</a>. TruSPAdes is a modification of genome assembler <a href="http://bioinf.spbau.ru/en/spades" target="_blank">SPAdes</a>.


<a name="sec1.1"></a>
<h3>1.1 TruSeq data</h3>
    TruSeq Synthetic Long Reads technology is based on fragmenting genomic DNA into large segments (about 10Kb long) and forming random pools of the resulting segments (each pool contains about 300 segments). Next, these fragments are clonally amplified, sheared, and marked with a unique barcode. Afterwards, they are sequenced using the standard Illumina short reads technology. All short reads originating from the same barcode are assembled together resulting in a set of long contigs (this step is called TruSeq barcode assembly). Ideally, the result of such sequencing effort for a single barcode is the collection of approximately 300 fragments (each fragment is about 10kb long) from a genome forming 300 long virtual reads. Together, these segments are expected to cover about 3 million nucleotides (barcode span).
    TruSPAdes is a tool for barcode assembly. Thus truSPAdes should be run separately on reads from each barcode. 
    
<a name="sec1.2"></a>
<h3>1.2 TruSPAdes performance</h3>
    TruSPAdes assembles average barcode (300000 reads) in 1 hour using one thread of Intel Xeon 2.27GHz processor and requires less than 2Gb RAM. TruSPAdes can be run in several threads to improve running time at the cost of increased memory consumption but we do not recommend it. Since truSPAdes should be run independantly on many barcodes we suggest running several instances of truSPAdes (on different barcodes) simultaneously each in one thread (or even on different nodes of cluster). This way standard TruSeq dataset with 394 barcodes can be processed in 24 hours using 16 threads and 32Gb RAM. Normally SPAdes assembles at least 2Mb virtual long reads from single barcode with N50 at least 7500.

<a name="sec2"></a>
<h2>2. Installation</h2>
<p>
    TruSPAdes requires a 64-bit Linux system or Mac OS and Python (supported versions are 2.4, 2.5, 2.6, 2.7, 3.2 and 3.3) to be pre-installed on it. To obtain SPAdes you can either download binaries or download source code and compile it yourself.

<a name="sec2.1"></a>
<h3>2.1 Downloading truSPAdes Linux binaries</h3>

<p>
    To download <a href="http://spades.bioinf.spbau.ru/truSPAdes-1.0/truSPAdes-1.0-Linux.tar.gz">truSPAdes Linux binaries</a> and extract them, go to the directory in which you wish truSPAdes to be installed and run:

<pre  class="code">
<code>
    wget http://spades.bioinf.spbau.ru/truSPAdes-1.0/truSPAdes-1.0-Linux.tar.gz
    tar -xzf truSPAdes-1.0-Linux.tar.gz
    cd SPAdes-1.0-Linux/bin/
</code>
</pre>

<p>
    In this case you do not need to run any installation scripts &ndash; truSPAdes is ready to use. The following files will be placed in the <code>bin</code> directory:
    <ul>
        <li><code>truspades.py</code>  (main executable script)</li>
        <li><code>truspades</code>  (assembly module)</li>
        <li><code>bwa-spades</code>  (<a href="http://bio-bwa.sourceforge.net" target="_blank">BWA</a> alignment module which is required for mismatch correction)</li>
    </ul>

<p>
    We also suggest adding truSPAdes installation directory to the <code>PATH</code> variable.
 
<a name="sec2.2"></a>
<h3>2.2 Downloading truSPAdes binaries for Mac</h3>

<p>
    To download <a href="http://spades.bioinf.spbau.ru/truSPAdes-1.0/truSPAdes-1.0-Darwin.tar.gz">truSPAdes binaries for Mac</a>, go to the directory in which you wish truSPAdes to be installed and run:

<pre  class="code">
<code>
    curl http://spades.bioinf.spbau.ru/truSPAdes-1.0/truSPAdes-1.0-Darwin.tar.gz -o truSPAdes-1.0-Darwin.tar.gz
    tar -zxf truSPAdes-1.0-Darwin.tar.gz
    cd truSPAdes-1.0-Darwin/bin/
</code>
</pre>

<p>
    Just as in Linux, SPAdes is ready to use and no further installation steps are required. You will get the same files in the <code>bin</code> directory:
    <ul>
        <li><code>truspades.py</code>  (main executable script)</li>
        <li><code>truspades</code>  (assembly module)</li>
        <li><code>bwa-spades</code>  (<a href="http://bio-bwa.sourceforge.net" target="_blank">BWA</a> alignment module which is required for mismatch correction)</li>
    </ul>

<p>
    We also suggest adding truSPAdes installation directory to the <code>PATH</code> variable.
 

<a name="sec2.3"></a>
<h3>2.3 Downloading and compiling truSPAdes source code</h3>
<p>
    If you wish to compile truSPAdes by yourself you will need the following libraries to be pre-installed:
    <ul>
        <li>g++ (version 4.7 or higher)</li>
        <li>cmake (version 2.8.8 or higher)</li>
        <li>zlib</li>
        <li>libbz2</li>
    </ul>

<p>
    If you meet these requirements, you can download the <a href="http://spades.bioinf.spbau.ru/truSPAdes-1.0/truSPAdes-1.0.tar.gz">truSPAdes source code</a>: 

<pre class="code">
<code>
    wget http://spades.bioinf.spbau.ru/truSPAdes-1.0/truSPAdes-1.0.tar.gz
    tar -xzf truSPAdes-1.0.tar.gz
    cd truSPAdes-1.0
</code>
</pre>

<p>
    and build it with the following script:

<pre class="code">
<code>
    ./truspades_compile.sh
</code>
</pre>

<p> 
    TruSPAdes will be built in the directory <code>./bin</code>. If you wish to install truSPAdes into another directory, you can specify full path of destination folder by running the following command in <code>bash</code> or <code>sh</code>:

<pre class="code">
<code>
    PREFIX=&lt;destination_dir> ./truspades_compile.sh
</code>
</pre>

<p>
    for example:

<pre class="code">
<code>
    PREFIX=/usr/local ./truspades_compile.sh
</code>
</pre>

<p>
   which will install truSPAdes into <code>/usr/local/bin</code>.
    
<p>
    After installation you will get the same files in <code>./bin</code> (or <code>&lt;destination_dir>/bin</code> if you specified PREFIX) directory:
    <ul>
        <li><code>truspades.py</code>  (main executable script)</li>
        <li><code>truspades</code>  (assembly module)</li>
        <li><code>bwa-spades</code>  (<a href="http://bio-bwa.sourceforge.net" target="_blank">BWA</a> alignment module which is required for mismatch correction)</li>
    </ul>

<p>
    We also suggest adding truSPAdes installation directory to the <code>PATH</code> variable.

<a name="sec2.4"></a>
<h3>2.4 Verifying your installation</h3>
<p>
    For testing purposes, truSPAdes comes with a toy data set. To try SPAdes on this data set, run:

<pre class="code">
<code>
    &lt;spades installation dir>/truspades.py --test
</code>
</pre>

<p>
    If you added truSPAdes installation directory to the <code>PATH</code> variable, you can run:

<pre class="code">
<code>
    truspades.py --test
</code>
</pre>

    For the simplicity we further assume that truSPAdes installation directory is added to the <code>PATH</code> variable.


<p>
    If the installation is successful, you will find the following information at the end of the log:

<pre class="code">
<code>
===== Assembling finished. 

 * Corrected reads are in truspades_test/corrected/
 * Assembled contigs are in truspades_test/contigs.fasta (contigs.fastg)
 * Assembled scaffolds are in truspades_test/scaffolds.fasta (scaffolds.fastg)

======= truSPAdes pipeline finished.

TruSPAdes log can be found here: &lt;your folder>/spades_test/spades.log

Thank you for using truSPAdes!

</code>
</pre>



<a name="sec3"></a>
<h2>3. Running truSPAdes</h2>

<a name="sec3.2"></a>
<h3>3.1 TruSPAdes command line options </h3>
<p>
    To run truSPAdes from the command line, type 

<pre class="code">
<code>
    spades.py [options] -1 &lt;left_reads_file> -2 &lt;right_reads_file> -o &lt;output_dir>
</code>
</pre>
Note that we assume that truSPAdes installation directory is added to the <code>PATH</code> variable (provide full path to truSPAdes executable otherwise: <code>&lt;truspades installation dir>/truspades.py</code>). 

<a name="basicopt"></a>
<h4>Basic options</h4>
<p>
    <code>-o &lt;output_dir> </code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Specify the output directory. Required option.
</p>

<p>
    <code>--test</code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Runs SPAdes on the toy data set; see <a href="#sec2.3">section 2.3</a>.
</p>

<p>
    <code>-h</code> (or <code>--help</code>)<br>
    &nbsp;&nbsp;&nbsp;&nbsp;Prints help.
</p>



<p>
    <code>--continue</code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Continues truSPAdes run from the specified output folder starting from the last available check-point. Check-points are made after:
    <ul>
        <li>iteration for each specified K value of assembly module is finished</li>
        <li>scaffold correction step is finished</li>
    </ul>
For example, if specified K values are 21, 33 and 55 and SPAdes was stopped or crashed during assembly stage with K = 55, you can run SPAdes with the <code>--continue</code> option specifying the same output directory. SPAdes will continue the run starting from the assembly stage with K = 55. Error correction module and iterations for K equal to 21 and 33 will not be run again.
Note that all options except <code>-o &lt;output_dir> </code> are ignored if <code>--continue</code> is set.
</p>

<p>
    <code>--restart-from &lt;check_point></code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Restart truSPAdes run from the specified output folder starting from the specified check-point. Check-points are:
    <ul>
        <li><code>as</code> &ndash; restart assembly module from the first iteration</li>
        <li><code>k&lt;int></code> &ndash; restart from the iteration with specified k values, e.g. k55</li>
        <li><code>sc</code> &ndash; restart scaffold correction</li>
        <li><code>pp</code> &ndash; restart postprocessing based on alignment</li>
    </ul>
In comparison to the <code>--continue</code> option, you can change some of the options when using <code>--restart-from</code>. You can change any option except all basic options. For example, if you ran assembler with k values 21,33,45,55, you can add one more iteration with k=77 with following options:<br>
     <code>--restart-from k55 -k 21,33,55,77 -o &lt;previous_output_dir></code>. 
<br>Since all files will be overwritten, do not forget to copy your assembly from the previous run if you need it.
</p>

<a name="inputdata"></a>
<h4>Input data</h4>
<p>
    <code>-1 &lt;file_name> </code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;File with forward reads.
</p>

<p>
    <code>-2 &lt;file_name> </code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;File with reverse reads.
</p>


<a name="advancedopt">
<h4>Advanced options</h4>

<p>
    <code>-t &lt;int></code> (or <code>--threads &lt;int></code>)<br>
    &nbsp;&nbsp;&nbsp;&nbsp;Number of threads. The default value is 1.
</p>

<p>
    <code>-m &lt;int></code> (or <code>--memory &lt;int></code>)<br>
    &nbsp;&nbsp;&nbsp;&nbsp;Set memory limit in Gb. TruSPAdes terminates if it reaches this limit. The default value is 250 Gb. Actual amount of consumed RAM will be below this limit. Make sure this value is correct for the given machine. TruSPAdes uses the limit value to automatically determine the sizes of various buffers, etc.
</p>

<p>
    <code>--tmp-dir &lt;dir_name></code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Set directory for temporary files from read error correction. The default value is <code>&lt;output_dir>/corrected/tmp</code>
</p>

<p>
    <code>-k &lt;int,int,...></code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Comma-separated list of k-mer sizes to be used (all values must be odd, less than 128 and listed in ascending order). The default value depends on the read length. E.g. for reads of length 100 <code>&lt;21,33,45,55</code> is used.
</p>

<a name="examples">
<h4>Examples</h4>
<p>
    To test the toy data set, you can also run the following command from the SPAdes <code>bin</code> directory:

<pre class="code">
<code>
    spades.py -1 ../share/truspades/test_dataset/ecoli_1K_1.fq.gz \
    -2 ../share/truspades/test_dataset/ecoli_1K_2.fq.gz -o truspades_test
</code>
</pre>

<p>
    If you want to use specific values of k:
<pre class="code">
<code>
    spades.py -1 left_reads.fastq -2 right_reads.fastq -k 21,31,41,51,61 -o output_folder
</code>
</pre>

<a name="sec3.5">
<h3>3.2 TruSPAdes output</h3>
<p>
    SPAdes stores all output files in <code>&lt;output_dir> </code>, which is set by the user. Resulting TruSeq long reads will be stored in <code>&lt;output_dir>/truseq_long_reads.fasta</code>
The full list of <code>&lt;output_dir></code> content is presented below: 
</p>
<pre>
    <code>truseq_long_reads.fasta</code> &ndash; <i>resulting truseq long reads</i>
    <code>params.txt</code> &ndash; <i>information about truSPAdes parameters in this run</i>
    <code>truspades.log</code> &ndash; <i>truSPAdes log</i>
    <code>dataset.info</code> &ndash; <i>internal configuration file</i>
    <code>input_dataset.yaml</code> &ndash; <i>internal YAML data set file</i>
    <code>K&lt;##>/</code> &ndash; <i>directory containing files from the run with K=&lt;##> </i>
    <code>SCC</code> &ndash; <i>directory containing files from the scaffold correction stage </i>
    <code>alignment/</code> &ndash; <i>directory containing alignment files used in postprocessing stage </i>
    <code>misc</code> &ndash; <i>directory containing scaffolds before postprocessing </i>
    <code>corrected</code> &ndash; <i>directory containing internal information </i>
</pre>

<p>
    TruSPAdes will overwrite these files and directories if they exist in the specified <code>&lt;output_dir></code>. 


<a name="sec3.6">
<h3>3.3 Assembly evaluation</h3>

<p>

    <a href="http://bioinf.spbau.ru/en/quast" target="_blank">QUAST</a> may be used to generate summary statistics (N50, maximum contig length, GC %, # genes found in a reference list or with built-in gene finding tools, etc.) for a single assembly. It may also be used to compare statistics for multiple assemblies of the same data set (e.g., truSPAdes run with different parameters, or several different assemblers).
<br>

<a name="sec5">
<h2>4. Feedback and bug reports</h2>
<p>
    Your comments, bug reports, and suggestions are very welcomed. They will help us to further improve truSPAdes.

<p>
    If you have any troubles running truSPAdes, please send us <code>params.txt</code> and <code>truspades.log</code> from the directory <code>&lt;output_dir></code>.

<p>
    Address for communications: <a href="mailto:spades.support@bioinf.spbau.ru" target="_blank">spades.support@bioinf.spbau.ru</a>.

<br/><br/><br/><br/><br/>

</body>
</html>
