<html>
<head>
    <title>truSPAdes 1.0 Manual</title>
    <style type="text/css">
        .code {
            background-color: lightgray;
        }
    </style>
</head>
<body>
<h1>TruSPAdes 1.0 Manual</h1>

1. <a href="#sec1">About truSPAdes</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;1.1. <a href="#sec1.1">TruSeq data</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;1.2. <a href="#sec1.2">TruSPAdes input</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;1.3. <a href="#sec1.3">TruSPAdes performance</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;1.4. <a href="#sec1.4">Installation</a><br>
2. <a href="#sec2">Running truSPAdes</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;3.1. <a href="#sec2.1">TruSPAdes command line options</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;3.2. <a href="#sec2.2">TruSPAdes output</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;3.3. <a href="#sec2.3">Assembly evaluation</a><br>
3. <a href="#sec3">Feedback and bug reports</a><br>
<br>

<a name="sec1"></a>
<h2>1. About truSPAdes</h2>
<p>
    TruSPAdes is an assembler for short reads produced by Illumina TruSeq Long Read technology. TruSPAdes accepts as an input a collection of demultiplexed TruSeq reads and assembles long virtual reads. This manual will help you to use truSPAdes.</br>

<a name="sec1.1"></a>
<h3>1.1 Illumina TruSeq data</h3>
    TruSeq Synthetic Long Reads technology is based on fragmenting genomic DNA into large segments (about 10Kb long) and forming random pools of the resulting segments (each pool contains about 300 segments). Next, these fragments are clonally amplified, sheared, and marked with a unique barcode. Afterwards, they are sequenced using the standard Illumina short reads technology. All short reads originating from the same barcode are assembled together resulting in a set of long contigs (this step is called TruSeq barcode assembly). Ideally, the result of such sequencing effort for a single barcode is the collection of 300 fragments (each fragment is about 10kb long) from a genome forming 300 long virtual reads. Together, these segments are expected to cover about 3 million nucleotides (barcode span).
    TruSPAdes is a tool for barcode assembly. TruSPAdes assembles each barcode separately and outputs assembled contigs as Virtual TruSeq Long Reads (TSLRs).

<a name="sec1.2"></a>
<h3>1.2 TruSPAdes input format</h3>
    TruSPades requires demultiplexed Illumina sequencing data as an input. All files with reads should be in the same directory and their names should follow the format described below.
        <l>
            <item>Names of files with left (right) reads should contain "R1" ("R2") as their substring.</item>
            <item>Names of files with paired-end reads that correspond to each other should differ only by substitution of "R1" to "R2"</item>
            <item>If TruSeq reads were obtained from several runs of Illumina sequencing machine (e.g. several lanes were produced from the same barcoding), reads from all runs should be put into the same directory and their names should contain "L<n>" as a substring (where n is the number of lane)</item>
            <item>Names of reads with the same barcode but from different lanes should </item>
        </l>

This naming format is normally used by Illumina .

<a name="sec1.3"></a>
<h3>1.2 TruSPAdes performance</h3>
    TruSPAdes assembles standard Illumina TruSeq dataset (consisting of 384 barcodes) in 30 hours using 8 threads of Intel Xeon 2.27GHz processor and requires less than 16 Gb RAM (2Gb RAM per thread).

<a name="sec1.4"></a>
<h2>1.3 Installation</h2>
<p>
     TruSPAdes comes as a part of SPAdes genome assembler. See <a href = manual.html#sec2>SPAdes manual</a> for installation instructions.

<a name="sec3"></a>
<h2>3. Running truSPAdes</h2>

<a name="sec3.2"></a>
<h3>3.1 TruSPAdes command line options </h3>
<p>
    To run truSPAdes from the command line, type 

<pre class="code">
<code>
    truspades.py [options] --input-dir &lt;left_reads_file> -2 &lt;right_reads_file> -o &lt;output_dir>
</code>
</pre>
Note that we assume that truSPAdes installation directory is added to the <code>PATH</code> variable (provide full path to truSPAdes executable otherwise: <code>&lt;truspades installation dir>/truspades.py</code>). 

<a name="basicopt"></a>
<h4>Basic options</h4>
<p>
    <code>-o &lt;output_dir> </code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Specify the output directory. Required option.
</p>

<p>
    <code>--test</code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Runs SPAdes on the toy data set; see <a href="#sec2.3">section 2.3</a>.
</p>

<p>
    <code>-h</code> (or <code>--help</code>)<br>
    &nbsp;&nbsp;&nbsp;&nbsp;Prints help.
</p>



<p>
    <code>--continue</code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Continues truSPAdes run from the specified output folder starting from the last available check-point. Check-points are made after:
    <ul>
        <li>iteration for each specified K value of assembly module is finished</li>
        <li>scaffold correction step is finished</li>
    </ul>
For example, if specified K values are 21, 33 and 55 and SPAdes was stopped or crashed during assembly stage with K = 55, you can run SPAdes with the <code>--continue</code> option specifying the same output directory. SPAdes will continue the run starting from the assembly stage with K = 55. Error correction module and iterations for K equal to 21 and 33 will not be run again.
Note that all options except <code>-o &lt;output_dir> </code> are ignored if <code>--continue</code> is set.
</p>

<p>
    <code>--restart-from &lt;check_point></code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Restart truSPAdes run from the specified output folder starting from the specified check-point. Check-points are:
    <ul>
        <li><code>as</code> &ndash; restart assembly module from the first iteration</li>
        <li><code>k&lt;int></code> &ndash; restart from the iteration with specified k values, e.g. k55</li>
        <li><code>sc</code> &ndash; restart scaffold correction</li>
        <li><code>pp</code> &ndash; restart postprocessing based on alignment</li>
    </ul>
In comparison to the <code>--continue</code> option, you can change some of the options when using <code>--restart-from</code>. You can change any option except all basic options. For example, if you ran assembler with k values 21,33,45,55, you can add one more iteration with k=77 with following options:<br>
     <code>--restart-from k55 -k 21,33,55,77 -o &lt;previous_output_dir></code>. 
<br>Since all files will be overwritten, do not forget to copy your assembly from the previous run if you need it.
</p>

<a name="inputdata"></a>
<h4>Input data</h4>
<p>
    <code>-1 &lt;file_name> </code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;File with forward reads.
</p>

<p>
    <code>-2 &lt;file_name> </code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;File with reverse reads.
</p>


<a name="advancedopt">
<h4>Advanced options</h4>

<p>
    <code>-t &lt;int></code> (or <code>--threads &lt;int></code>)<br>
    &nbsp;&nbsp;&nbsp;&nbsp;Number of threads. The default value is 1.
</p>

<p>
    <code>-m &lt;int></code> (or <code>--memory &lt;int></code>)<br>
    &nbsp;&nbsp;&nbsp;&nbsp;Set memory limit in Gb. TruSPAdes terminates if it reaches this limit. The default value is 250 Gb. Actual amount of consumed RAM will be below this limit. Make sure this value is correct for the given machine. TruSPAdes uses the limit value to automatically determine the sizes of various buffers, etc.
</p>

<p>
    <code>--tmp-dir &lt;dir_name></code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Set directory for temporary files from read error correction. The default value is <code>&lt;output_dir>/corrected/tmp</code>
</p>

<p>
    <code>-k &lt;int,int,...></code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Comma-separated list of k-mer sizes to be used (all values must be odd, less than 128 and listed in ascending order). The default value depends on the read length. E.g. for reads of length 100 <code>&lt;21,33,45,55</code> is used.
</p>

<a name="examples">
<h4>Examples</h4>
<p>
    To test the toy data set, you can also run the following command from the SPAdes <code>bin</code> directory:

<pre class="code">
<code>
    spades.py -1 ../share/truspades/test_dataset/ecoli_1K_1.fq.gz \
    -2 ../share/truspades/test_dataset/ecoli_1K_2.fq.gz -o truspades_test
</code>
</pre>

<p>
    If you want to use specific values of k:
<pre class="code">
<code>
    spades.py -1 left_reads.fastq -2 right_reads.fastq -k 21,31,41,51,61 -o output_folder
</code>
</pre>

<a name="sec3.5">
<h3>3.2 TruSPAdes output</h3>
<p>
    SPAdes stores all output files in <code>&lt;output_dir> </code>, which is set by the user. Resulting TruSeq long reads will be stored in <code>&lt;output_dir>/truseq_long_reads.fasta</code>
The full list of <code>&lt;output_dir></code> content is presented below: 
</p>
<pre>
    <code>truseq_long_reads.fasta</code> &ndash; <i>resulting truseq long reads</i>
    <code>params.txt</code> &ndash; <i>information about truSPAdes parameters in this run</i>
    <code>truspades.log</code> &ndash; <i>truSPAdes log</i>
    <code>dataset.info</code> &ndash; <i>internal configuration file</i>
    <code>input_dataset.yaml</code> &ndash; <i>internal YAML data set file</i>
    <code>K&lt;##>/</code> &ndash; <i>directory containing files from the run with K=&lt;##> </i>
    <code>SCC</code> &ndash; <i>directory containing files from the scaffold correction stage </i>
    <code>alignment/</code> &ndash; <i>directory containing alignment files used in postprocessing stage </i>
    <code>misc</code> &ndash; <i>directory containing scaffolds before postprocessing </i>
    <code>corrected</code> &ndash; <i>directory containing internal information </i>
</pre>

<p>
    TruSPAdes will overwrite these files and directories if they exist in the specified <code>&lt;output_dir></code>. 


<a name="sec3.6">
<h3>3.3 Assembly evaluation</h3>

<p>

    <a href="http://bioinf.spbau.ru/en/quast" target="_blank">QUAST</a> may be used to generate summary statistics (N50, maximum contig length, GC %, # genes found in a reference list or with built-in gene finding tools, etc.) for a single assembly. It may also be used to compare statistics for multiple assemblies of the same data set (e.g., truSPAdes run with different parameters, or several different assemblers).
<br>

<a name="sec5">
<h2>4. Feedback and bug reports</h2>
<p>
    Your comments, bug reports, and suggestions are very welcomed. They will help us to further improve truSPAdes.

<p>
    If you have any troubles running truSPAdes, please send us <code>params.txt</code> and <code>truspades.log</code> from the directory <code>&lt;output_dir></code>.

<p>
    Address for communications: <a href="mailto:spades.support@bioinf.spbau.ru" target="_blank">spades.support@bioinf.spbau.ru</a>.

<br/><br/><br/><br/><br/>

</body>
</html>
