\documentclass[17pt]{extarticle} % another possible option: 20 pt
\usepackage{ifpdf}
\ifpdf
\usepackage{cmap}
\fi
\usepackage[english]{babel}
\usepackage{palatino}
\usepackage[x11names,svgnames]{xcolor}
\usepackage{amsmath,amssymb,amsfonts}
\usepackage{tikz}
\usepackage{amsthm}
%\usepackage{pgf}
%\usepackage{multicol}
\usepackage{xspace}
\usepackage{graphicx}
%\usepackage{multicol}
%\usepackage[section]{algorithm} % [section] is use to define the numbering mode
%\usepackage{algorithmic} 
\usepackage[a1paper,left=2.5cm,right=.5cm,top=2.5cm,bottom=.5cm,foot=0cm]{geometry}
\usepackage{poster}
\usepackage{booktabs}
\usepackage{multirow,multicol}
\usepackage{simplewick}

\usetikzlibrary{positioning,fadings,fit}

%\usepackage{pgfplots}

%\theoremstyle{definition}
%\newtheorem{definition}{Definition}[section]
%\newtheorem{proposition}{Proposition}[section]
%\newtheorem{corollary}{Corollary}[section]
%\newtheorem{theorem}{Theorem}[section]


\def\myarc#1#2#3#4{
  \draw[color=#1,line width=2pt,->] (360*#2/18:#4mm) arc (360*#2/18:360*#3/18:#4mm);
}%#1 -- color, #2 -- from, #3 -- to, #4 -- radius (mm)

\def\ra{red}
\def\rb{green!50!black}
\def\rc{cyan}
\def\rd{magenta}
\def\re{blue}
\def\rf{orange}

\newcommand{\myignore}[1]{}

\begin{document}
\pagestyle{empty}

\begin{center}
\begin{tikzpicture}

%%%%%%%%%%%%%%%GRID
%\draw[step=1cm,help lines] (0,0) grid (54,78);
%\foreach \x in {0,...,54} { \node at (\x,-1) {$\x$}; \node at (\x,79) {$\x$}; }
%\foreach \x in {0,...,78} { \node at (-1,\x) {$\x$}; \node at (55,\x) {$\x$}; }

%%%%%%%%%%%%%%%TITLE
\begin{scope}
\draw (0,78) [color=SteelBlue3,line width=2mm,rounded corners=.5cm,fill=blue!10!white] rectangle (54,68);
\node at (27,76) {\color{OrangeRed} \fontsize{60pt}{1em} \bf Earmark graph approach to \emph{de~novo} genome assembly};
\node[rectangle,text width=16cm,text centered,anchor=north west] at (1,74) {
  {\bf \color{DodgerBlue3} \fontsize{32pt}{1em}\selectfont Mikhail~E.~Dvorkin}
};
\node[rectangle,text width=16cm,text centered,anchor=north west] at (18,74) {
  {\bf \color{DodgerBlue3} \fontsize{32pt}{1em}\selectfont Alexander~S.~Kulikov}
};
\node[rectangle,text width=16cm,text centered,anchor=north west] at (36,74) {
  {\bf \color{DodgerBlue3} \fontsize{32pt}{1em}\selectfont Max~A.~Alekseyev}
};
\node[rectangle,text width=52cm,text centered,anchor=north west] at (1,72.5) {
  \fontsize{26pt}{1em}\selectfont
  Algorithmic Biology Lab, Academic University of the Russian Academy of Sciences, St.~Petersburg, Russia\\
  {\tt http://bioinf.spbau.ru/en/}
\\[+5mm]
% We've got to include this:
\small{The presentation of this work was made possible by a Travel Fellowship awarded by ISCB
with grant funds obtained from the Department of Energy Office of 
Science,\\
the National Science Foundation Bio-Directorate,
and the NIH National Institute of General Medical Sciences.}
};

\end{scope}

%%%%%%%%%%%%%%%PROBLEM STATEMENT AND ABSTRACT
\begin{scope}
\draw ( 0,65) [color=SteelBlue3,line width=2mm,rounded corners=.5cm] rectangle (26,58);
\draw (28,65) [color=SteelBlue3,line width=2mm,rounded corners=.5cm] rectangle (54,58);
\node[draw=SteelBlue3,line width=2mm,rounded corners=.5cm,anchor=south west,fill=blue!10!white,inner sep=3mm] at (28,65) {\fontsize{32pt}{1em} 
\bf Typical genome assembly setting\phantom{y}};
\node[draw=SteelBlue3,line width=2mm,rounded corners=.5cm,anchor=south west,fill=blue!10!white,inner sep=3mm] at (0,65) {\fontsize{32pt}{1em} 
\bf Abstract\phantom{y}};
\node[rectangle,anchor=west,text width=25cm,inner sep=5mm] at (0,61.5) {
  A common approach to assembling a genome from short reads is constructing
  the de~Bruijn graph on all $k$-mers from the given set of reads and finding
  a traversal of edges in this graph. We propose a new approach that allows
  to decrease the graph size without losing the essential information from
  the input data. Instead of using all the $k$-mers from a read we take only
  a few of them (call them {\it earmarks}). Besides an obvious advantage
  of requiring less memory and time for constructing, the resulting earmark
  graph has several other advantages over the de~Bruijn graph.
};
\node[rectangle,anchor=west,text width=25cm,inner sep=5mm] at (28,61.5) {
  %\fontsize{26pt}{2em}\selectfont
  \Large
  {\color{OrangeRed}\bf Input:} a set of substrings (called reads) 
    ${\cal R} \subseteq \{{\tt A,C,G,T}\}^r$ of an unknown circular string 
    $S \in \{{\tt A,C,G,T}\}^*$ (called genome).\\[+3mm]
  {\color{OrangeRed}\bf Output:} the genome $S$.\\[+3mm]
  %~\\
  {\color{OrangeRed}\bf Complications: } reads may contain errors.
};
%at (27,64) {\fontsize{26pt}{1em} Given a set of reads ${\cal R} \subseteq \{{\tt 
%A,C,G,T}\}^r$ of a genome $S \in \{{\tt A,C,G,T}\}^*$, find the genome $S$.};
\end{scope}

%%%%%%%%%%%%%%%EXAMPLE
\begin{scope}[yshift=-7cm]
  \draw (0,62) [color=SteelBlue3,line width=2mm,rounded corners=.5cm] rectangle (54,32);
  \node[draw=SteelBlue3,line width=2mm,rounded corners=.5cm,anchor=south west,fill=blue!10!white,inner sep=3mm] at (0,62) {\fontsize{32pt}{1em} \bf 
    Standard and new approaches: an example\phantom{y}};

  \node[rectangle,draw=SteelBlue3,line width=3pt,fill=white,anchor=north west,text width=8cm,rounded corners=.5cm,inner sep=3mm] at (0.5,61.5) {\fontsize{26pt}{1em} \bf In this example:\\ ~~$r=8$\\ ~~$k=4$};

  %%% READS
  \node[rectangle,line width=2pt,text width=5cm,text centered,anchor=west] (reads) at (1,48.5) {
    {\bf set ${\cal R}$ of reads of length $r$}\\
    {\color{\ra}  $R_1$: $\overrightarrow{\tt CATGTAGT}$}\\
    {\color{\rb}  $R_2$: $\overrightarrow{\tt GTAACATG}$}\\
    {\color{\rc}  $R_3$: $\overrightarrow{\tt AGTGTACA}$}\\
    {\color{\rd}  $R_4$: $\overrightarrow{\tt TGTAGTGT}$}\\
    {\color{\re}  $R_5$: $\overrightarrow{\tt CATGTAAC}$}\\
    {\color{\rf}  $R_6$: $\overrightarrow{\tt TAACATGT}$}\\
  };

  \begin{scope}[xshift=48.5cm,yshift=48.5cm,scale=1.2,transform shape]
    \node (genometop)  at (0,3.5) {};
    \node (genomedown) at (0,-3.5) {};
    \node (genomeleft) at (-3.5,0) {};

    \foreach \i/\a in {0/C,1/A,2/T,3/G,4/T,5/A,6/A,7/C,8/A,9/T,10/G,11/T,12/A,13/G,14/T,15/G,16/T,17/A} {
      \node[fill=white] at (360*\i/18:2cm) {${\tt \a}$};
      %\node at (360*\i/18:1.5cm) {{\small \i}};
    }
    \node at (0,0) {genome $S$};
    %\draw[color=red,->] (360*5/18:2.2cm) arc (360*5/18:360*11/18:2.2cm);
    %\myarc{black}{1}{18}{16}
    \myarc{\ra}{7}{14}{24}  %R1
    \myarc{\rb}{3}{10}{26}  %R2
    \myarc{\rc}{-6}{1}{28}  %R3
    \myarc{\rd}{9}{16}{30}  %R4
    \myarc{\re}{0}{7}{32}   %R5
    \myarc{\rf}{3}{10}{34}  %R6
  \end{scope}

  %\draw [->,line width=2pt] (reads) to[out=90,in=180] node[sloped,fill=white] {de~Bruijn graph} (23,53);
  %\draw [->,line width=2pt] (reads) to[out=-90,in=180] node[sloped,fill=white] {earmark graph} (23,44);


  \begin{scope}[scale=1.3,transform shape,line width=2pt]
    \begin{scope}[xshift=18cm,yshift=40cm]
      \foreach \x/\y/\t in {0.5/4.5/CATG, 2.5/3.5/ATGT, 4/3/TGTA, 5.5/3/GTAG, 5.5/1.5/GTAA,
        7/4/GTAC, 7/3/TAGT, 7/0.5/TAAC, 9/3/AGTG, 9/0.5/AACA, 10.5/4/TACA, 10.5/2.5/GTGT, 12/4/ACAT}
      \node[draw,rectangle,scale=0.7] (\t) at (\x,\y) {${\tt \t}$};

      \foreach \s/\t in {TAGT/AGTG, AGTG/GTGT, TGTA/GTAA, GTAC/TACA, TACA/ACAT}
        \draw [->] (\s) to (\t);
      \draw [->] (GTGT) to[bend left=20] (TGTA);
      \draw [->] (TGTA) to[bend left=20] (GTAC);
      
      \foreach \s/\t in {CATG/ATGT,ATGT/TGTA,TGTA/GTAG,GTAG/TAGT}
        \draw [->,color=\ra] (\s) to (\t);
      \foreach \s/\t in {GTAA/TAAC, TAAC/AACA}
        \draw [->,color=\rb] (\s) to (\t);
      \draw [->,color=\rb] (AACA) to[bend right=45] (ACAT);
      \draw [->,color=\rb] (ACAT) to[bend right=20] (CATG);

      \node[rectangle,fit=(CATG) (ACAT) (AACA)] (debruijngraph) {};
    \end{scope}

    \begin{scope}[xshift=18cm,yshift=29cm]
      \foreach \x/\y/\t in {0.5/4.5/CATG, 5.5/1.5/GTAA,
        7/3/TAGT, 9/3/AGTG, 10.5/4/TACA, 12/4/ACAT}
      \node[draw,rectangle,scale=0.7] (\t) at (\x,\y) {${\tt \t}$};

      \foreach \x/\y/\t in {2.5/3.5/ATGT, 4/3/TGTA, 5.5/3/GTAG, 
        7/4/GTAC, 7/0.5/TAAC, 9/0.5/AACA, 10.5/2.5/GTGT}
      \node[draw=black!30,rectangle,scale=0.7] (\t) at (\x,\y) {${\tt \textcolor{black!30} \t}$};

      \foreach \s/\t in {TAGT/AGTG, AGTG/TACA}
        \draw [->] (\s) to (\t);
      \draw [->] (CATG) to[bend right=20] (GTAA);
      \draw [->,color=\rb] (GTAA) to[bend right=40,in=-90] (ACAT);
      \draw [->,color=\rb] (ACAT) to[bend right=20] (CATG);
      \draw [->,color=\ra] (CATG) to[bend left=5] (TAGT);
      \draw [->] (TACA) to[bend right=15] (CATG);

      \node[rectangle,fit=(CATG) (ACAT) (AACA)] (earmarkgraph) {};
    \end{scope}
  \end{scope}

  %\node[anchor=south west,scale=0.5] (debruijngraph) at (30,51) {\includegraphics{de.pdf}};
  %\node[anchor=south west,scale=0.5] (earmarkgraph)  at (30,37) {\includegraphics{en.pdf}};

  \node[rectangle,line width=2pt,text width=3cm,text centered,anchor=west] (earmarks) at (9,38) {
    {\bf earmarks}\\
    ${\tt CATG}$ ${\tt TACA}$\\
    ${\tt GTAA}$ ${\tt ACAT}$\\
    ${\tt AGTG}$ ${\tt TAGT}$
  };

  \node[rectangle,line width=2pt,text width=5cm,text centered,anchor=west] (earmarkedreads) at (15,38) {
    {\color{\ra} $R_1: \contraction{}{\tt C}{\tt AT}{\tt G}%
    \contraction{\tt CATG}{\tt T}{\tt AG}{\tt T}%
    {\tt CATGTAGT}$ }

    {\color{\rb} $R_2: \contraction{}{\tt G}{\tt TA}{\tt A}%
    \contraction{\tt GTA}{\tt A}{\tt CA}{\tt T}%
    \bcontraction{\tt GTAA}{\tt C}{\tt AT}{\tt G}%
    {\tt GTAACATG}$ }

    {\dots}
  };

  \begin{scope}[draw=SteelBlue3,line width=3pt,fill=white]
    \draw [->] (reads) to[out=90,in=180] node[fill=white,rectangle,text width=6cm] 
      {\color{DodgerBlue3} 1. represent each read as a path on its $r-k+1$ consecutive \mbox{$k$-mers}} (debruijngraph);

    \draw [->] (debruijngraph) to[out=15,in=90] node[fill=white,sloped,rectangle,text width=7cm] 
      {\color{DodgerBlue3} 2. simplify the graph (compress simple paths, clip tips, remove erroneous connections and bulges) and find a cycle containing all the edges} (genometop);

    \draw [->] (reads) to[out=-90,in=180] node[fill=white,sloped,rectangle,text width=5cm] 
      {\color{DodgerBlue3} 1. construct a set ${\cal E} \subseteq \{{\tt A,C,G,T}\}^k$ of earmarked \mbox{$k$-mers} such that each read contains at least
      two earmarked $k$-mers} (earmarks);

    \draw [->] (earmarks) to[out=-90,in=-90] node[fill=white,rectangle,text width=6cm] 
      {\color{DodgerBlue3} 2. find the earmarks in the reads} (earmarkedreads);

    \draw [->] (earmarkedreads) to[out=-45,in=-135] node[fill=white,rectangle,text width=5cm] 
      {\color{DodgerBlue3} 3. represent each read as a path on its consecutive earmarks} (earmarkgraph);

    \draw [->] (earmarkgraph) to[out=-30,in=-90] node[fill=white,sloped,rectangle,text width=7cm] 
      {\color{DodgerBlue3} 4. simplify the graph (compress simple paths, clip tips, remove erroneous connections and bulges) and find a cycle containing all the edges} (genomedown);
  \end{scope}

  \node at (32,61) {\fontsize{26pt}{1em} \color{OrangeRed} \bf (standard) de~Bruijn graph approach};
  \node at (33,33) {\fontsize{26pt}{1em} \color{OrangeRed} \bf (new) earmark graph approach};
  %\draw[color=OrangeRed,dashed,line width=5pt,->] (7,48) node[above] {genome assembly problem} -- (46,48);
  \path (reads) edge[color=OrangeRed,dashed,line width=5pt,->] node[above] {\fontsize{26pt}{1em} \color{OrangeRed} \bf \emph{de~novo} genome assembly problem} (genomeleft);
\end{scope}

%%%%%%%%%%%%%%%BENCHMARKING
\begin{scope}
  \draw (0,22) [color=SteelBlue3,line width=2mm,rounded corners=.5cm] rectangle (22,0);
  \node[draw=SteelBlue3,line width=2mm,rounded corners=.5cm,anchor=south west,fill=blue!10!white,inner sep=3mm] at (0,22) {\fontsize{32pt}{1em} \bf 
    Benchmarking\phantom{y}};
  
  \node[rectangle,text width=22cm,text centered,anchor=west] at (0,11) {
    \input{../data/stat.tex}
    ~\\
    ~\\
    \begin{center}
    E. coli genome,\\
    reads of length $r = 100$,\\
    filtered with Quake;\\
    $k = 25$
    \end{center}
  };
\end{scope}

%%%%%%%%%%%%%%%ADVANTAGES
\begin{scope}
  \draw (24,22) [color=SteelBlue3,line width=2mm,rounded corners=.5cm] rectangle (54,8.5);
  \node[draw=SteelBlue3,line width=2mm,rounded corners=.5cm,anchor=south west,fill=blue!10!white,inner sep=3mm] at (24,22) {\fontsize{32pt}{1em} \bf 
    Advantages of earmark graph over de~Bruijn graph\phantom{y}};
  \node[rectangle,text width=28cm,anchor=north west] at (25,21.5) {
    %\fontsize{24pt}{2em}\selectfont
    \large
    \begin{description}
    \item[\color{OrangeRed} \bf Smaller size.]
    The earmark graph requires less time and memory for construction.
    Note also that one can control the size of the earmark graph 
    by varying the size of the set of earmarks. Also, it is easy to see 
    that in an extreme case when ${\cal E}$ is just the set of all the 
    $k$-mers of the input reads, the earmark graph coincides with the
    de~Bruijn graph.
    \item[\color{OrangeRed} \bf Simpler structure.]
    In the example above, the de~Bruijn graph contains two vertices $v$ with 
    $\textrm{indegree}(v)\times\textrm{outdegree}(v) \ge 2$, while the 
    earmark graph contains only one such vertex. 
    This corresponds to a simpler representation of repeats 
    in the genome and simplifies the problem of finding a cycle in the 
    graph.
    %At the same time, the 
    %genome can be spelled through both graphs.
    %To give an example, consider two reads ${\tt TTGCAC}$ and ${\tt 
    %ATGCAT}$ sharing a $4$-mer ${\tt TGCA}$. They are represented as two paths, shown 
    %in Fig.~\ref{fig:repa}, in the de~Bruijn graph built on $3$-mers
    %(this is a part of the de~Bruijn graph from Fig.~\ref{fig:debruijn}).
    %This is a typical repeat. The edge ${\tt TGC}\rightarrow{\tt GCA}$ has two
    %incoming edges and two outgoing edges. While spelling a genome
    %through this part of the graph it is not clear which incoming edge 
    %corresponds to which outgoing edge. However in the earmarked graph
    %this repeat may be already resolved if the $3$-mers {\tt TGC} and {\tt GCA}
    %are not earmarked, see Fig.~\ref{fig:repb}.
    \item[\color{OrangeRed} \bf Using trusted $k$-mers.] 
    By restricting earmarks to $k$-mers present in many reads (``trusted'' $k$-mers)
    one can reduce the number of erroneous edges resulting from 
    sequencing errors in the reads.
    \end{description}
  };
\end{scope}

%%%%%%%%%%%%%%%REFERENCES
\begin{scope}
  \draw (24,6) [color=SteelBlue3,line width=2mm,rounded corners=.5cm] rectangle (54,0);
  \node[draw=SteelBlue3,line width=2mm,rounded corners=.5cm,anchor=south west,fill=blue!10!white,inner sep=3mm] at (24,6) {\fontsize{32pt}{1em} \bf References\phantom{y}};
  \node[rectangle,text width=28cm,anchor=north west] at (24.5,5.5) {\small
%    1. A Bruijn\\
%    2. De Bruijn\\
%    3. Minimizers
\newcommand{\etalchar}[1]{$^{#1}$}
%\begin{thebibliography}{RHH{\etalchar{+}}04}
\begin{enumerate}
\item 
Pevzner, Tang, Tesler.
\newblock {\em De novo} repeat classification and fragment assembly.
\newblock {\em RECOMB}, 2004.%, pages 213--222.

%\bibitem[PTW01]{PW01}
\item 
Pevzner, Tang, Waterman.
\newblock An Eulerian path approach to DNA fragment assembly.
\newblock {\em Proc Natl Acad Sci USA}, 2001. %98(17):9748--9753, 2001.

%\bibitem[RHH{\etalchar{+}}04]{RW04}
\item 
Roberts, Hayes, Hunt, Mount, Yorke.
\newblock Reducing storage requirements for biological sequence comparison.
\newblock {\em Bioinformatics}, 2004. %20:3363--3369, December 2004.

%\bibitem[ZB08]{Z08}
\item 
Zerbino, Birney.
\newblock Velvet: algorithms for {\em de novo} short read assembly using de Bruijn graphs.
\newblock {\em Genome Res.}, 2008. %18(5):821--9, 2008.
\end{enumerate}
%\end{thebibliography}

  };
\end{scope}

\end{tikzpicture}
\end{center}


\end{document}
